{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Probe Designer for Isoforms\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basci env\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import time\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "\n",
    "# data process of file from ncbi\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "\n",
    "# get gene data from ncbi\n",
    "from Bio import Entrez\n",
    "\n",
    "# blast and xml file process\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "# add package to sys var\n",
    "# os.chdir(os.path.dirname(os.path.abspath(__file__)))\n",
    "try:\n",
    "    sys.path.append(\"../lib\")\n",
    "except:\n",
    "    pass\n",
    "\n",
    "# dir\n",
    "project_name = \"2023.11.20_Isoform_trials\"\n",
    "dataset_path = \"/home/akikawa/repos/probe_designer/dataset/\"\n",
    "\n",
    "os.makedirs(f\"{dataset_path}/{project_name}\", exist_ok=True)\n",
    "os.chdir(f\"{dataset_path}/{project_name}\")\n",
    "\n",
    "current_time = time.localtime()\n",
    "formatted_time = time.strftime(\"%Y%m%d_%H%M%S\", current_time)\n",
    "tmp = f\"./trials/{formatted_time}/\"\n",
    "output = f\"./trials/{formatted_time}/\"\n",
    "pre_binding_dir = f\"{tmp}/pre_binding/\"\n",
    "os.makedirs(tmp, exist_ok=True)\n",
    "\n",
    "# basic variables\n",
    "gene_name_list_tosearch = \"gene_name_list_tosearch.txt\"\n",
    "pre_binding_file_suffix = \"_pre_binding.fasta\"\n",
    "total_pre_binding_file_name = \"_total_pre_binding.fasta\"\n",
    "\n",
    "# tmp file\n",
    "gene_name_list_file = \"1_gene_name_list.txt\"\n",
    "gene_id_name_file = \"2_id_list.txt\"\n",
    "gene_seq_in_file = \"3_gene_seq_in_file.gb\"\n",
    "pre_binding_num_file = \"4_pre_binding_num.json\"\n",
    "blast_results_file = \"5_blast_results.xml\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "\n",
    "workdir = r'./dataset/2023.11.22_Isoform_trials/'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get seq file of each gene from ensembl dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Gene_BRCA1: 100%|██████████| 56/56 [00:24<00:00,  2.27it/s]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "import requests\n",
    "\n",
    "species = \"human\"\n",
    "gene = \"BRCA1\"\n",
    "seq_type = \"cds\"\n",
    "\n",
    "def get_seqs(gene, species, seq_type):\n",
    "    lookup_url = f\"http://rest.ensembl.org/lookup/symbol/{species}/{gene}?content-type=application/json\"\n",
    "    gene_id = requests.get(url=lookup_url).json()[\"id\"]\n",
    "\n",
    "    transcripts_url = f\"http://rest.ensembl.org/overlap/id/{gene_id}?feature=transcript;content-type=application/json\"\n",
    "    transcripts = requests.get(url=transcripts_url).json()\n",
    "\n",
    "    # Get sequences for each transcript\n",
    "    sequences = {}\n",
    "    for transcript in tqdm(transcripts, desc=f'Gene_{gene}'):\n",
    "        try:\n",
    "            seq_name = \"_\".join([transcript[\"id\"], transcript[\"external_name\"], transcript['biotype']])\n",
    "            seq_url = f\"http://rest.ensembl.org/sequence/id/{transcript['id']}?type={seq_type};content-type=application/json\"\n",
    "            seq_response = requests.get(seq_url).json()\n",
    "            sequences[seq_name] = seq_response[\"seq\"]\n",
    "        except: continue\n",
    "\n",
    "    return sequences\n",
    "\n",
    "sequences = get_seqs(gene, species, seq_type)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene ENST00000461221_BRCA1-204_nonsense_mediated_decay: \tNo valid positions, please loosen your threshold conditions.\n",
      "Gene ENST00000700183_BRCA1-237_nonsense_mediated_decay: \tNo valid positions, please loosen your threshold conditions.\n",
      "Gene ENST00000492859_BRCA1-218_nonsense_mediated_decay: \tNo valid positions, please loosen your threshold conditions.\n",
      "Gene ENST00000461798_BRCA1-206_nonsense_mediated_decay: \tNo valid positions, please loosen your threshold conditions.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ENST00000497488_BRCA1-222_protein_coding: 100%|██████████| 42/42 [00:16<00:00,  2.53it/s]\n",
      "ENST00000489037_BRCA1-216_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.54it/s]\n",
      "ENST00000478531_BRCA1-214_protein_coding: 100%|██████████| 24/24 [00:09<00:00,  2.50it/s]\n",
      "ENST00000357654_BRCA1-203_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.54it/s]\n",
      "ENST00000473961_BRCA1-211_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.53it/s]\n",
      "ENST00000477152_BRCA1-213_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.53it/s]\n",
      "ENST00000352993_BRCA1-201_protein_coding: 100%|██████████| 24/24 [00:09<00:00,  2.49it/s]\n",
      "ENST00000493919_BRCA1-220_protein_coding: 100%|██████████| 24/24 [00:09<00:00,  2.49it/s]\n",
      "ENST00000494123_BRCA1-221_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.54it/s]\n",
      "ENST00000471181_BRCA1-209_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.54it/s]\n",
      "ENST00000652672_BRCA1-232_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.53it/s]\n",
      "ENST00000634433_BRCA1-228_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.53it/s]\n",
      "ENST00000476777_BRCA1-212_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.54it/s]\n",
      "ENST00000470026_BRCA1-208_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.54it/s]\n",
      "ENST00000618469_BRCA1-226_protein_coding: 100%|██████████| 43/43 [00:17<00:00,  2.52it/s]\n",
      "ENST00000461574_BRCA1-205_protein_coding: 100%|██████████| 43/43 [00:17<00:00,  2.53it/s]\n",
      "ENST00000644555_BRCA1-231_protein_coding: 100%|██████████| 24/24 [00:09<00:00,  2.49it/s]\n",
      "ENST00000468300_BRCA1-207_protein_coding: 100%|██████████| 23/23 [00:09<00:00,  2.49it/s]\n",
      "ENST00000644379_BRCA1-230_protein_coding: 100%|██████████| 43/43 [00:17<00:00,  2.49it/s]\n",
      "ENST00000484087_BRCA1-215_protein_coding: 100%|██████████| 24/24 [00:09<00:00,  2.48it/s]\n",
      "ENST00000586385_BRCA1-223_protein_coding: 100%|██████████| 50/50 [00:19<00:00,  2.54it/s]\n",
      "ENST00000591534_BRCA1-224_protein_coding: 100%|██████████| 68/68 [00:26<00:00,  2.59it/s]\n",
      "ENST00000591849_BRCA1-225_protein_coding: 100%|██████████| 113/113 [00:41<00:00,  2.71it/s]\n",
      "ENST00000493795_BRCA1-219_protein_coding: 100%|██████████| 43/43 [00:16<00:00,  2.53it/s]\n",
      "ENST00000461221_BRCA1-204_nonsense_mediated_decay: 0it [00:00, ?it/s]\n",
      "ENST00000491747_BRCA1-217_protein_coding: 100%|██████████| 24/24 [00:09<00:00,  2.48it/s]\n",
      "ENST00000700182_BRCA1-236_protein_coding: 100%|██████████| 20/20 [00:08<00:00,  2.47it/s]\n",
      "ENST00000700183_BRCA1-237_nonsense_mediated_decay: 0it [00:00, ?it/s]\n",
      "ENST00000492859_BRCA1-218_nonsense_mediated_decay: 0it [00:00, ?it/s]\n",
      "ENST00000642945_BRCA1-229_nonsense_mediated_decay: 100%|██████████| 11/11 [00:04<00:00,  2.46it/s]\n",
      "ENST00000461798_BRCA1-206_nonsense_mediated_decay: 0it [00:00, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "from lib.search_binding import step_by_step\n",
    "\n",
    "\n",
    "def calculate_optimal_alignment_similarity(sub1, sub2):\n",
    "    max_similarity = 0\n",
    "    for shift in range(-len(sub1) + 1, len(sub1)):\n",
    "        shifted_similarity = sum((sub1[i] == sub2[i - shift]) for i in range(len(sub1)) if 0 <= i - shift < len(sub2))\n",
    "        max_similarity = max(max_similarity, shifted_similarity)\n",
    "    return max_similarity\n",
    "\n",
    "def find_most_distinctive_substrings(sequences_dict, substring_length=40, top_n=10):\n",
    "    if any(len(seq) < substring_length for seq in sequences_dict.values()):\n",
    "        raise ValueError(\"One or more sequences are shorter than the specified substring length.\")\n",
    "    \n",
    "    all_substrings = {}\n",
    "    for name, seq in sequences_dict.items():\n",
    "        _, _, selected_substrings, _ = step_by_step(\n",
    "        seq,\n",
    "        BDS_len=40,\n",
    "        BDS_num=50,\n",
    "        min_gap=1,\n",
    "        better_gap=40,\n",
    "        gene=name,\n",
    "        G_min=0.25,\n",
    "        G_max=0.7,\n",
    "        G_consecutive=5,\n",
    "        Tm_low=50,\n",
    "        Tm_high=65,\n",
    "        show_process=False, \n",
    "        warn=False\n",
    "        )\n",
    "        \n",
    "        all_substrings[name] = selected_substrings\n",
    "\n",
    "    top_substrings = {}\n",
    "    for name, substrings in all_substrings.items():\n",
    "        distinctiveness_scores = {}\n",
    "        for sub in tqdm(substrings, desc=name):\n",
    "            distinctiveness_scores[sub] = min(calculate_optimal_alignment_similarity(sub, other_sub)\n",
    "                                              for other_name, other_substrings in all_substrings.items()\n",
    "                                              if other_name != name\n",
    "                                              for other_sub in other_substrings)\n",
    "\n",
    "        top_n_substrings = sorted(distinctiveness_scores, key=distinctiveness_scores.get)[:top_n]\n",
    "        top_substrings[name] = top_n_substrings\n",
    "\n",
    "    return top_substrings\n",
    "\n",
    "distinctive_substrings = find_most_distinctive_substrings(sequences, substring_length=40)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENST00000497488_BRCA1-222_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000489037_BRCA1-216_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000478531_BRCA1-214_protein_coding': ['AGCGTCTCTGAAGACTGCTCAGGGCTATCCTCTCAGAGTG',\n",
       "  'AGAAGGCCTTTCTGCTGACAAGTTTGAGGTGTCTGCAGAT',\n",
       "  'GTTGTTGATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGC',\n",
       "  'GATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACG',\n",
       "  'TGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGAT',\n",
       "  'GTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGATT',\n",
       "  'GATCTAGAGGGAACCCCTTACCTGGAATCTGGAATCAGCC',\n",
       "  'ATCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGG',\n",
       "  'TCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGC',\n",
       "  'CCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCA'],\n",
       " 'ENST00000357654_BRCA1-203_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000473961_BRCA1-211_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000477152_BRCA1-213_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000352993_BRCA1-201_protein_coding': ['AGCGTCTCTGAAGACTGCTCAGGGCTATCCTCTCAGAGTG',\n",
       "  'AGAAGGCCTTTCTGCTGACAAGTTTGAGGTGTCTGCAGAT',\n",
       "  'GTTGTTGATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGC',\n",
       "  'GATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACG',\n",
       "  'TGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGAT',\n",
       "  'GTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGATT',\n",
       "  'GATCTAGAGGGAACCCCTTACCTGGAATCTGGAATCAGCC',\n",
       "  'ATCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGG',\n",
       "  'TCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGC',\n",
       "  'CCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCA'],\n",
       " 'ENST00000493919_BRCA1-220_protein_coding': ['AGCGTCTCTGAAGACTGCTCAGGGCTATCCTCTCAGAGTG',\n",
       "  'AGAAGGCCTTTCTGCTGACAAGTTTGAGGTGTCTGCAGAT',\n",
       "  'GTTGTTGATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGC',\n",
       "  'GATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACG',\n",
       "  'TGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGAT',\n",
       "  'GTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGATT',\n",
       "  'GATCTAGAGGGAACCCCTTACCTGGAATCTGGAATCAGCC',\n",
       "  'ATCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGG',\n",
       "  'TCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGC',\n",
       "  'CCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCA'],\n",
       " 'ENST00000494123_BRCA1-221_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000471181_BRCA1-209_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000652672_BRCA1-232_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000634433_BRCA1-228_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000476777_BRCA1-212_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000470026_BRCA1-208_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000618469_BRCA1-226_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000461574_BRCA1-205_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000644555_BRCA1-231_protein_coding': ['AGCGTCTCTGAAGACTGCTCAGGGCTATCCTCTCAGAGTG',\n",
       "  'AGAAGGCCTTTCTGCTGACAAGTTTGAGGTGTCTGCAGAT',\n",
       "  'GTTGTTGATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGC',\n",
       "  'GATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACG',\n",
       "  'TGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGAT',\n",
       "  'GTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGATT',\n",
       "  'GATCTAGAGGGAACCCCTTACCTGGAATCTGGAATCAGCC',\n",
       "  'ATCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGG',\n",
       "  'TCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGC',\n",
       "  'CCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCA'],\n",
       " 'ENST00000468300_BRCA1-207_protein_coding': ['AGCGTCTCTGAAGACTGCTCAGGGCTATCCTCTCAGAGTG',\n",
       "  'AGAAGGCCTTTCTGCTGACAAGTTTGAGGTGTCTGCAGAT',\n",
       "  'GTTGTTGATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGC',\n",
       "  'GATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACG',\n",
       "  'TGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGAT',\n",
       "  'GTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGATT',\n",
       "  'GATCTAGAGGGAACCCCTTACCTGGAATCTGGAATCAGCC',\n",
       "  'ATCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGG',\n",
       "  'TCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGC',\n",
       "  'CCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCA'],\n",
       " 'ENST00000644379_BRCA1-230_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000484087_BRCA1-215_protein_coding': ['AGCGTCTCTGAAGACTGCTCAGGGCTATCCTCTCAGAGTG',\n",
       "  'AGAAGGCCTTTCTGCTGACAAGTTTGAGGTGTCTGCAGAT',\n",
       "  'GTTGTTGATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGC',\n",
       "  'GATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACG',\n",
       "  'TGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGAT',\n",
       "  'GTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGATT',\n",
       "  'GATCTAGAGGGAACCCCTTACCTGGAATCTGGAATCAGCC',\n",
       "  'ATCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGG',\n",
       "  'TCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGC',\n",
       "  'CCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCA'],\n",
       " 'ENST00000586385_BRCA1-223_protein_coding': ['GTGTGGTGCTTCTGTGGTGAAGGAGCTTTCATCATTCACC',\n",
       "  'GTGGTGCTTCTGTGGTGAAGGAGCTTTCATCATTCACCCT',\n",
       "  'GGTGCTTCTGTGGTGAAGGAGCTTTCATCATTCACCCTTG',\n",
       "  'TGCTTCTGTGGTGAAGGAGCTTTCATCATTCACCCTTGGC',\n",
       "  'TCTGTGGTGAAGGAGCTTTCATCATTCACCCTTGGCACAG',\n",
       "  'GGTGTCCACCCAATTGTGGTTGTGCAGCCAGATGCCTGGA',\n",
       "  'TTGTGGTTGTGCAGCCAGATGCCTGGACAGAGGACAATGG',\n",
       "  'TGCAGCCAGATGCCTGGACAGAGGACAATGGCTTCCATGC',\n",
       "  'CAGCCAGATGCCTGGACAGAGGACAATGGCTTCCATGCAA',\n",
       "  'AGGACAATGGCTTCCATGCAATTGGGCAGATGTGTGAGGC'],\n",
       " 'ENST00000591534_BRCA1-224_protein_coding': ['GATCTAGAGGGAACCCCTTACCTGGAATCTGGAATCAGCC',\n",
       "  'ATCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGG',\n",
       "  'TCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGC',\n",
       "  'CCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCA',\n",
       "  'CTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCAA',\n",
       "  'TTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCAAC',\n",
       "  'TCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCAACA',\n",
       "  'CTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCAACAT',\n",
       "  'TGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCAACATA',\n",
       "  'AAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCAACATACC'],\n",
       " 'ENST00000591849_BRCA1-225_protein_coding': ['GTGTGGTGCTTCTGTGGTGAAGGAGCTTTCATCATTCACC',\n",
       "  'TGTGGTGCTTCTGTGGTGAAGGAGCTTTCATCATTCACCC',\n",
       "  'GTGGTGCTTCTGTGGTGAAGGAGCTTTCATCATTCACCCT',\n",
       "  'TGGTGCTTCTGTGGTGAAGGAGCTTTCATCATTCACCCTT',\n",
       "  'GGTGCTTCTGTGGTGAAGGAGCTTTCATCATTCACCCTTG',\n",
       "  'GTGCTTCTGTGGTGAAGGAGCTTTCATCATTCACCCTTGG',\n",
       "  'TGCTTCTGTGGTGAAGGAGCTTTCATCATTCACCCTTGGC',\n",
       "  'GCTTCTGTGGTGAAGGAGCTTTCATCATTCACCCTTGGCA',\n",
       "  'TCTGTGGTGAAGGAGCTTTCATCATTCACCCTTGGCACAG',\n",
       "  'GGTGTCCACCCAATTGTGGTTGTGCAGCCAGATGCCTGGA'],\n",
       " 'ENST00000493795_BRCA1-219_protein_coding': ['AGGCTGAGGAGGAAGTCTTCTACCAGGCATATTCATGCGC',\n",
       "  'GGAGGCTCTAGGTTTTGTCTATCATCTCAGTTCAGAGGCA',\n",
       "  'ACAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCT',\n",
       "  'CAGCCTATGGGAAGTAGTCATGCATCTCAGGTTTGTTCTG',\n",
       "  'GCGTCCAGAAAGGAGAGCTTAGCAGGAGTCCTAGCCCTTT',\n",
       "  'TTCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGG',\n",
       "  'TCACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGC',\n",
       "  'CACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCC',\n",
       "  'ACCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCA',\n",
       "  'CCCATACACATTTGGCTCAGGGTTACCGAAGAGGGGCCAA'],\n",
       " 'ENST00000461221_BRCA1-204_nonsense_mediated_decay': [],\n",
       " 'ENST00000491747_BRCA1-217_protein_coding': ['AGCGTCTCTGAAGACTGCTCAGGGCTATCCTCTCAGAGTG',\n",
       "  'AGAAGGCCTTTCTGCTGACAAGTTTGAGGTGTCTGCAGAT',\n",
       "  'GTTGTTGATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGC',\n",
       "  'GATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACG',\n",
       "  'TGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGAT',\n",
       "  'GTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGATT',\n",
       "  'GATCTAGAGGGAACCCCTTACCTGGAATCTGGAATCAGCC',\n",
       "  'ATCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGG',\n",
       "  'TCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGC',\n",
       "  'CCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCA'],\n",
       " 'ENST00000700182_BRCA1-236_protein_coding': ['AGCGTCTCTGAAGACTGCTCAGGGCTATCCTCTCAGAGTG',\n",
       "  'AGAAGGCCTTTCTGCTGACAAGTTTGAGGTGTCTGCAGAT',\n",
       "  'GTTGTTGATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGC',\n",
       "  'GATGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACG',\n",
       "  'TGTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGAT',\n",
       "  'GTGGAGGAGCAACAGCTGGAAGAGTCTGGGCCACACGATT',\n",
       "  'GATCTAGAGGGAACCCCTTACCTGGAATCTGGAATCAGCC',\n",
       "  'ATCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGG',\n",
       "  'TCCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGC',\n",
       "  'CCTTCTGAAGACAGAGCCCCAGAGTCAGCTCGTGTTGGCA'],\n",
       " 'ENST00000700183_BRCA1-237_nonsense_mediated_decay': [],\n",
       " 'ENST00000492859_BRCA1-218_nonsense_mediated_decay': [],\n",
       " 'ENST00000642945_BRCA1-229_nonsense_mediated_decay': ['GCAAACAGTGTCTCCCTCTGACACCCAGGCTGGAGTTCAG',\n",
       "  'AAACAGTGTCTCCCTCTGACACCCAGGCTGGAGTTCAGTG',\n",
       "  'AACAGTGTCTCCCTCTGACACCCAGGCTGGAGTTCAGTGG',\n",
       "  'ACAGTGTCTCCCTCTGACACCCAGGCTGGAGTTCAGTGGT',\n",
       "  'CAGTGTCTCCCTCTGACACCCAGGCTGGAGTTCAGTGGTG',\n",
       "  'AGTGTCTCCCTCTGACACCCAGGCTGGAGTTCAGTGGTGT',\n",
       "  'GTGTCTCCCTCTGACACCCAGGCTGGAGTTCAGTGGTGTG',\n",
       "  'TGTCTCCCTCTGACACCCAGGCTGGAGTTCAGTGGTGTGA',\n",
       "  'GTCTCCCTCTGACACCCAGGCTGGAGTTCAGTGGTGTGAT',\n",
       "  'TCTCCCTCTGACACCCAGGCTGGAGTTCAGTGGTGTGATC'],\n",
       " 'ENST00000461798_BRCA1-206_nonsense_mediated_decay': []}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "distinctive_substrings"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Binding site Searcher\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "position_searching_CTRB1: 100%|██████████| 656/656 [00:00<00:00, 11751.96it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "condition for gene CTRB1 too harsh, loose to get better results\n",
      "[2, 12, 21, 33, 42, 51, 67, 119, 129, 138, 151, 160, 173, 185, 194, 214, 234, 243, 252, 265, 274, 285, 294, 303, 313, 323, 333, 342, 353, 362, 371, 422, 433, 442, 451, 460, 484, 499, 508, 519, 528, 537, 546, 555, 566, 575, 586, 595, 604, 623, 639, 648]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "from lib.search_binding import step_by_step, find_max_min_difference_fixed_length_subsequence, gb_extract\n",
    "\n",
    "# Initiation of array\n",
    "binding_site_FOIs = [\n",
    "    \"accession\",\n",
    "    \"gene_name\",\n",
    "    \"mol_type\",\n",
    "    \"organism\",\n",
    "    \"pos_on_seq\",\n",
    "    \"binding\",\n",
    "    \"Tm_l\",\n",
    "    \"Tm_r\",\n",
    "    \"wanted\",\n",
    "]\n",
    "align_FOIs = [\"align_num\", \"align_accession\", \"align_descrip\", \"plus/minus\"]\n",
    "FOI = pd.DataFrame(columns=binding_site_FOIs + align_FOIs)\n",
    "\n",
    "# Search binding sites on mRNA sequence\n",
    "file_in = tmp + gene_seq_in_file\n",
    "file_out_dir = pre_binding_dir\n",
    "try:\n",
    "    os.mkdir(file_out_dir)\n",
    "except:\n",
    "    pass\n",
    "\n",
    "pre_binding_num = {}\n",
    "\n",
    "# initialization of file\n",
    "with open(file_out_dir + total_pre_binding_file_name, \"w\") as handle:\n",
    "    handle.write(\"\")\n",
    "\n",
    "for gene_seq_gb in SeqIO.parse(tmp + gene_seq_in_file, \"genbank\"):\n",
    "    id, gene_name, mol_type, organism, seq = gb_extract(gene_seq_gb)\n",
    "\n",
    "    Tm_l, Tm_r, selected_substrings, pos_on_seq = step_by_step(\n",
    "        seq,\n",
    "        BDS_len=40,\n",
    "        BDS_num=50,\n",
    "        min_gap=1,\n",
    "        better_gap=40,\n",
    "        gene=gene_name,\n",
    "        G_min=0.25,\n",
    "        G_max=0.7,\n",
    "        G_consecutive=5,\n",
    "        Tm_low=50,\n",
    "        Tm_high=65,\n",
    "    )\n",
    "    \n",
    "    record_list = []\n",
    "    for i, pre_binding_tmp in enumerate(selected_substrings):\n",
    "        record_list.append(\n",
    "            SeqRecord(\n",
    "                Seq(pre_binding_tmp),\n",
    "                id=\"pre_binding\" + str(i),\n",
    "                description=\"|\".join([id, gene_name, organism, mol_type]),\n",
    "            )\n",
    "        )\n",
    "\n",
    "    # add information about binding sites to FOI\n",
    "    add = pd.DataFrame(\n",
    "        {\n",
    "            \"accession\": [id] * len(selected_substrings),\n",
    "            \"gene_name\": [gene_name] * len(selected_substrings),\n",
    "            \"mol_type\": [mol_type] * len(selected_substrings),\n",
    "            \"organism\": [organism] * len(selected_substrings),\n",
    "            \"binding\": selected_substrings,\n",
    "            \"Tm_l\": Tm_l,\n",
    "            \"Tm_r\": Tm_r,\n",
    "            \"pos_on_seq\": pos_on_seq,\n",
    "        }\n",
    "    )\n",
    "    FOI = pd.concat([FOI, add], ignore_index=True)\n",
    "\n",
    "    file_out = file_out_dir + gene_name + pre_binding_file_suffix\n",
    "    # write pre_binding to files\n",
    "    with open(file_out, \"w\") as f:\n",
    "        for new_record in record_list:\n",
    "            SeqIO.write(new_record, f, \"fasta\")\n",
    "    with open(file_out_dir + total_pre_binding_file_name, \"a\") as handle:\n",
    "        for new_record in record_list:\n",
    "            SeqIO.write(new_record, handle, \"fasta\")\n",
    "\n",
    "    # record the num of pre_binding for each gene\n",
    "    pre_binding_num[f\"{id}_{gene_name}\"] = len(selected_substrings)\n",
    "\n",
    "with open(tmp + pre_binding_num_file, \"w\") as f:\n",
    "    json.dump(pre_binding_num, f)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blast and extract blast results\n",
    "\n",
    "NCBIXML: https://homolog.us/Biopython/Bio.Blast.NCBIXML.html#read/0\n",
    "\n",
    "BlastRecord: https://biopython.org/docs/1.75/api/Bio.Blast.Record.html\n",
    "\n",
    "XMLReader: https://codebeautify.org/xmlviewer#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(file_out_dir + total_pre_binding_file_name, \"r\") as f:\n",
    "    fasta_string = f.read()\n",
    "txid = [2697049]  # organism\n",
    "\n",
    "# Submit BLAST search and get handle object\n",
    "handle = NCBIWWW.qblast(\n",
    "    program=\"blastn\",\n",
    "    megablast=\"yes\",\n",
    "    database=\"refseq_rna\",\n",
    "    sequence=fasta_string,\n",
    "    url_base=\"https://blast.ncbi.nlm.nih.gov/Blast.cgi\",\n",
    "    format_object=\"Alignment\",\n",
    "    format_type=\"Xml\",\n",
    ")\n",
    "\n",
    "# read handle object and save to a file\n",
    "with open(tmp + blast_results_file, \"w\") as f:\n",
    "    f.write(handle.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract interested information from blast_results\n",
    "align_num = []\n",
    "\n",
    "# read the id/plus-minus part/align_num\n",
    "with open(tmp + blast_results_file, \"r\") as blast_output:\n",
    "    blast_records = NCBIXML.parse(blast_output)\n",
    "    loca = 0\n",
    "    for blast_record in blast_records:\n",
    "        align_accession = []\n",
    "        align_descrip_list = []\n",
    "        # get align num of each binding site\n",
    "        length = len(blast_record.alignments)\n",
    "        align_num.append(length)\n",
    "        for i in range(length):\n",
    "            descrip = blast_record.descriptions[i].title.split(\"|\")\n",
    "            # get accession and descrip of each align seq\n",
    "            align_accession.append(descrip[3])\n",
    "            align_descrip_list.append(descrip[-1])\n",
    "        FOI.loc[loca, \"align_accession\"] = \"|\".join(str(_) for _ in align_accession)\n",
    "\n",
    "        # add align_descrip to df\n",
    "        FOI.loc[loca, \"align_descrip\"] = \"|\".join(str(_) for _ in align_descrip_list)\n",
    "\n",
    "        # get plus/minus of each align seq\n",
    "        p_m = [blast_record.alignments[_].hsps[0].frame[1] for _ in range(length)]\n",
    "\n",
    "        # add plus/minus to df\n",
    "        try:\n",
    "            FOI.loc[loca, \"plus/minus\"] = \",\".join([str(_) for _ in p_m])\n",
    "        except:\n",
    "            FOI.loc[loca, \"plus/minus\"] = \"NAN\"\n",
    "\n",
    "        loca += 1\n",
    "\n",
    "FOI[\"align_num\"] = align_num"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select wanted binding site\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "FOI[\"wanted\"] = [True] * len(FOI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "condition for gene CTRB1 too harsh, loose to get better results\n",
      "[89, 99, 108, 120, 129]\n"
     ]
    }
   ],
   "source": [
    "# sieve for the suitable binding site\n",
    "gene_name_list = [_.upper() for _ in gene_name_list]\n",
    "gene_name_list_out = [i for i in gene_name_list]\n",
    "for i in range(len(FOI)):\n",
    "    # check gene_name\n",
    "    gene_name = FOI.loc[i, \"gene_name\"]\n",
    "    if gene_name.upper() not in gene_name_list:\n",
    "        FOI.loc[i, \"wanted\"] = False\n",
    "    else:\n",
    "        try:\n",
    "            gene_name_list_out.remove(gene_name)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # check DNA or mRNA type\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        if FOI.loc[i, \"mol_type\"] != \"mRNA\":\n",
    "            FOI.loc[i, \"wanted\"] = False\n",
    "\n",
    "    # check gene_organism name\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        spe_ori, gene_ori = FOI.loc[i, \"organism\"], FOI.loc[i, \"gene_name\"]\n",
    "        descrip = FOI.loc[i, \"align_descrip\"].split(\"|\")\n",
    "        for des in descrip:\n",
    "            if gene_ori not in des and spe_ori in des:\n",
    "                FOI.loc[i, \"wanted\"] = False\n",
    "                break\n",
    "\n",
    "    # check plus/minus\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        if pd.isnull(FOI.loc[i, \"plus/minus\"]):\n",
    "            FOI.loc[i, \"wanted\"] = False\n",
    "        else:\n",
    "            pm_list = FOI.loc[i, \"plus/minus\"].split(\",\")\n",
    "            if \"-1\" not in pm_list:\n",
    "                FOI.loc[i, \"wanted\"] = False\n",
    "\n",
    "# write the whole information of interest to a excel file in tmp dir\n",
    "FOI.to_excel(tmp + \"probes_sieve.xlsx\")\n",
    "\n",
    "out_tmp = FOI[FOI[\"wanted\"] == True]\n",
    "output_df = pd.DataFrame()\n",
    "for gene in out_tmp.gene_name.unique():\n",
    "    pos_of_True = list(out_tmp[out_tmp.gene_name == gene][\"pos_on_seq\"])\n",
    "    best_pos = find_max_min_difference_fixed_length_subsequence(\n",
    "        pos_of_True,\n",
    "        length=3,\n",
    "        min_gap=40,\n",
    "        better_gap=80,\n",
    "        gene=gene,\n",
    "    )\n",
    "    out_subset = out_tmp[out_tmp.gene_name == gene]\n",
    "    out_subset = out_subset[out_subset[\"pos_on_seq\"].isin(best_pos)]\n",
    "    output_df = pd.concat([output_df, out_subset])\n",
    "\n",
    "# write the output to a xlsx file\n",
    "output_df.to_excel(output + \"probes_wanted.xlsx\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
