{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Probe Designer\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basci env\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import time\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "\n",
    "# data process of file from ncbi\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "\n",
    "# # get gene data from ncbi\n",
    "# from Bio import Entrez\n",
    "\n",
    "# # blast and xml file process\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "# add package to sys var\n",
    "# os.chdir(os.path.dirname(os.path.abspath(__file__)))\n",
    "# sys.path.append(\"../lib\")\n",
    "\n",
    "# dir\n",
    "workdir = './dataset/2023.12.15_Sindy_Isoforms/'\n",
    "os.makedirs(workdir, exist_ok=True)\n",
    "\n",
    "current_time = time.localtime()\n",
    "formatted_time = time.strftime(\"%Y%m%d_%H%M%S\", current_time)\n",
    "\n",
    "output = os.path.join(workdir, 'results', formatted_time)\n",
    "pre_binding_dir = os.path.join(output, \"pre_binding\")\n",
    "os.makedirs(output, exist_ok=True)\n",
    "\n",
    "# basic variables\n",
    "gene_name_list_tosearch = \"gene_name_list_tosearch.txt\"\n",
    "pre_binding_file_suffix = \"_pre_binding.fasta\"\n",
    "total_pre_binding_file_name = \"_total.fasta\"\n",
    "\n",
    "# tmp file\n",
    "gene_name_list_file = \"gene_list.xlsx\"\n",
    "pre_binding_num_file = \"pre_binding_num.json\"\n",
    "blast_results_file = \"blast_results.xml\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "organism = 'mouse'\n",
    "gene_info = pd.read_excel(os.path.join(workdir, gene_name_list_file), sheet_name='Sheet1')\n",
    "gene_list = list(gene_info['gene_name'].unique())\n",
    "id_list = list(gene_info['ensembl_id'])\n",
    "isoform_list = list(gene_info['isoform'])\n",
    "\n",
    "sequences_of_all = dict()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get seq from ensembl dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Gene:\tGm16024-201: 100%|██████████| 7/7 [00:05<00:00,  1.23it/s]\n",
      "Gene:\tGm47551-201: 100%|██████████| 10/10 [00:07<00:00,  1.26it/s]\n",
      "Gene:\tGm26822-201: 100%|██████████| 1/1 [00:00<00:00,  1.93it/s]\n",
      "Gene:\tChek2-202: 100%|██████████| 3/3 [00:03<00:00,  1.07s/it]\n",
      "Gene:\tMalat1-204: 100%|██████████| 1/1 [00:00<?, ?it/s]\n",
      "Gene:\tBrms1l-204: 100%|██████████| 6/6 [00:04<00:00,  1.25it/s]\n",
      "Gene:\tGlipr1-205: 100%|██████████| 4/4 [00:02<00:00,  1.52it/s]\n",
      "Gene:\tMalat1-203: 100%|██████████| 17/17 [00:10<00:00,  1.60it/s]\n",
      "Gene:\tGm11032-201: 100%|██████████| 1/1 [00:00<?, ?it/s]\n",
      "Gene:\tGm48610-201: 100%|██████████| 1/1 [00:00<00:00,  1.57it/s]\n",
      "Gene:\tTbk1-202: 100%|██████████| 4/4 [00:02<00:00,  1.91it/s]\n",
      "Gene:\tCep57-202: 100%|██████████| 11/11 [00:07<00:00,  1.38it/s]\n",
      "Gene:\tCisd3-204: 100%|██████████| 4/4 [00:05<00:00,  1.29s/it]\n",
      "Gene:\tGm47777-201: 100%|██████████| 1/1 [00:00<00:00,  2.00it/s]\n",
      "Gene:\tMalat1-201: 100%|██████████| 23/23 [00:15<00:00,  1.51it/s]\n",
      "Gene:\tChurc1-202: 100%|██████████| 6/6 [00:03<00:00,  1.85it/s]\n",
      "Gene:\tGm43994-201: 100%|██████████| 1/1 [00:01<00:00,  1.50s/it]\n",
      "Gene:\tChurc1-203: 100%|██████████| 6/6 [00:05<00:00,  1.04it/s]\n",
      "Gene:\tUchl5-208: 100%|██████████| 6/6 [00:04<00:00,  1.29it/s]\n",
      "Gene:\tGtpbp4-203: 100%|██████████| 4/4 [00:02<00:00,  1.71it/s]\n",
      "Gene:\tGm45892-201: 100%|██████████| 7/7 [00:04<00:00,  1.43it/s]\n",
      "Gene:\tYthdc2-204: 100%|██████████| 3/3 [00:02<00:00,  1.14it/s]\n",
      "Gene:\tRab34-207: 100%|██████████| 7/7 [00:05<00:00,  1.34it/s]\n",
      "Gene:\tGm10563-201: 100%|██████████| 1/1 [00:00<00:00, 1003.90it/s]\n",
      "Gene:\tGm12976-202: 100%|██████████| 2/2 [00:01<00:00,  1.31it/s]\n",
      "Gene:\tRpn2-211: 100%|██████████| 7/7 [00:04<00:00,  1.62it/s]\n",
      "Gene:\tGm47408-201: 100%|██████████| 4/4 [00:01<00:00,  2.05it/s]\n",
      "Gene:\tDock10-206: 100%|██████████| 6/6 [00:03<00:00,  1.53it/s]\n",
      "Gene:\tGm28625-201: 100%|██████████| 1/1 [00:00<00:00,  1.70it/s]\n",
      "Gene:\tMterf3-203: 100%|██████████| 10/10 [00:07<00:00,  1.28it/s]\n",
      "Gene:\tWdr31-206: 100%|██████████| 9/9 [00:06<00:00,  1.47it/s]\n",
      "Gene:\tGm47106-201: 100%|██████████| 10/10 [00:07<00:00,  1.32it/s]\n",
      "Gene:\tPign-204: 100%|██████████| 5/5 [00:05<00:00,  1.00s/it]\n",
      "Gene:\tSnf8-204: 100%|██████████| 4/4 [00:02<00:00,  1.93it/s]\n",
      "Gene:\tTnfrsf14-205: 100%|██████████| 5/5 [00:04<00:00,  1.20it/s]\n",
      "Gene:\tChchd1-204: 100%|██████████| 5/5 [00:02<00:00,  1.82it/s]\n",
      "Gene:\tCaprin2-208: 100%|██████████| 6/6 [00:05<00:00,  1.14it/s]\n",
      "Gene:\tGm35082-202: 100%|██████████| 2/2 [00:01<00:00,  1.98it/s]\n",
      "Gene:\t4930572G02Rik-201: 100%|██████████| 5/5 [00:04<00:00,  1.22it/s]\n",
      "Gene:\tMrpl48-203: 100%|██████████| 5/5 [00:03<00:00,  1.50it/s]\n",
      "Gene:\tGm27000-201: 100%|██████████| 3/3 [00:01<00:00,  2.00it/s]\n",
      "Gene:\tR3hdm1-211: 100%|██████████| 12/12 [00:08<00:00,  1.44it/s]\n",
      "Gene:\tAC158554.1-201: 100%|██████████| 8/8 [00:05<00:00,  1.47it/s]\n",
      "Gene:\tRbm39-205: 100%|██████████| 17/17 [00:13<00:00,  1.22it/s]\n",
      "Gene:\tPecam1-208: 100%|██████████| 6/6 [00:04<00:00,  1.24it/s]\n",
      "Gene:\tGm47403-201: 100%|██████████| 1/1 [00:01<00:00,  1.52s/it]\n",
      "Gene:\tGm17555-201: 100%|██████████| 1/1 [00:00<?, ?it/s]\n",
      "Gene:\tGm12953-201: 100%|██████████| 3/3 [00:01<00:00,  1.83it/s]\n",
      "Gene:\tFgd4-205: 100%|██████████| 13/13 [00:10<00:00,  1.27it/s]\n",
      "Gene:\tAY512931-202: 100%|██████████| 2/2 [00:00<00:00,  2.05it/s]\n",
      "Gene:\tTpm3-215: 100%|██████████| 11/11 [00:06<00:00,  1.76it/s]\n",
      "Gene:\tSord-201: 100%|██████████| 1/1 [00:01<00:00,  1.15s/it]\n",
      "Gene:\tPak2-202: 100%|██████████| 2/2 [00:01<00:00,  1.76it/s]\n",
      "Gene:\tZfp229-204: 100%|██████████| 6/6 [00:05<00:00,  1.04it/s]\n",
      "Gene:\tVamp3-203: 100%|██████████| 4/4 [00:03<00:00,  1.20it/s]\n",
      "Gene:\tAC125141.1-201: 100%|██████████| 2/2 [00:01<00:00,  1.32it/s]\n",
      "Gene:\tGm21750-201: 100%|██████████| 1/1 [00:00<00:00, 1001.51it/s]\n",
      "Gene:\tSsbp2-203: 100%|██████████| 8/8 [00:05<00:00,  1.40it/s]\n",
      "Gene:\tUnc13d-205: 100%|██████████| 8/8 [00:05<00:00,  1.44it/s]\n",
      "Gene:\tGm21917-201: 100%|██████████| 4/4 [00:02<00:00,  1.91it/s]\n",
      "Gene:\tRnf214-202: 100%|██████████| 8/8 [00:08<00:00,  1.09s/it]\n",
      "Gene:\tTcf7l2-204: 100%|██████████| 23/23 [00:14<00:00,  1.58it/s]\n",
      "Gene:\tThg1l-202: 100%|██████████| 5/5 [00:05<00:00,  1.07s/it]\n",
      "Gene:\tGm48278-201: 100%|██████████| 3/3 [00:01<00:00,  2.00it/s]\n",
      "Gene:\tDnajc14-205: 100%|██████████| 5/5 [00:03<00:00,  1.27it/s]\n",
      "Gene:\tRipor1-201: 100%|██████████| 3/3 [00:01<00:00,  1.80it/s]\n",
      "Gene:\tDcbld2-203: 100%|██████████| 7/7 [00:05<00:00,  1.24it/s]\n",
      "Gene:\tGm47013-201: 100%|██████████| 1/1 [00:00<00:00,  1.98it/s]\n",
      "Gene:\tTubb3-201: 100%|██████████| 2/2 [00:01<00:00,  1.91it/s]\n",
      "Gene:\tAdd2-206: 100%|██████████| 12/12 [00:07<00:00,  1.53it/s]\n",
      "Gene:\tSpock2-201: 100%|██████████| 4/4 [00:03<00:00,  1.01it/s]\n",
      "Gene:\tRnasek-203: 100%|██████████| 5/5 [00:04<00:00,  1.22it/s]\n",
      "Gene:\tKndc1-202: 100%|██████████| 2/2 [00:01<00:00,  1.87it/s]\n",
      "Gene:\tCoro7-202: 100%|██████████| 7/7 [00:06<00:00,  1.06it/s]\n",
      "Gene:\tGak-201: 100%|██████████| 14/14 [00:07<00:00,  1.88it/s]\n",
      "Gene:\tRrp1-206: 100%|██████████| 9/9 [00:06<00:00,  1.47it/s]\n",
      "Gene:\tUchl1-202: 100%|██████████| 3/3 [00:02<00:00,  1.34it/s]\n",
      "Gene:\tRbck1-203: 100%|██████████| 7/7 [00:06<00:00,  1.12it/s]\n",
      "Gene:\tPacsin1-207: 100%|██████████| 10/10 [00:08<00:00,  1.20it/s]\n",
      "Gene:\tDhx16-202: 100%|██████████| 7/7 [00:04<00:00,  1.43it/s]\n",
      "Gene:\tActl6b-205: 100%|██████████| 9/9 [00:05<00:00,  1.74it/s]\n",
      "Gene:\tActb-201: 100%|██████████| 9/9 [00:08<00:00,  1.06it/s]\n",
      "Gene:\tRbck1-206: 100%|██████████| 5/5 [00:03<00:00,  1.39it/s]\n",
      "Gene:\tDennd6b-201: 100%|██████████| 8/8 [00:06<00:00,  1.18it/s]\n",
      "Gene:\tGm11373-201: 100%|██████████| 1/1 [00:00<00:00,  2.08it/s]\n",
      "Gene:\tCox8a-201: 100%|██████████| 2/2 [00:01<00:00,  1.85it/s]\n",
      "Gene:\tPhf24-202: 100%|██████████| 9/9 [00:06<00:00,  1.32it/s]\n",
      "Gene:\tCnnm1-204: 100%|██████████| 3/3 [00:01<00:00,  1.92it/s]\n",
      "Gene:\tLy6h-201: 100%|██████████| 8/8 [00:05<00:00,  1.51it/s]\n",
      "Gene:\tCamk2b-209: 100%|██████████| 15/15 [00:12<00:00,  1.21it/s]\n",
      "Gene:\tGramd1b-206: 100%|██████████| 18/18 [00:14<00:00,  1.26it/s]\n",
      "Gene:\tCamkv-202: 100%|██████████| 6/6 [00:03<00:00,  1.98it/s]\n",
      "Gene:\tA230050P20Rik-204: 100%|██████████| 4/4 [00:03<00:00,  1.31it/s]\n",
      "Gene:\tGrk2-204: 100%|██████████| 6/6 [00:04<00:00,  1.29it/s]\n",
      "Gene:\tAtxn2l-206: 100%|██████████| 6/6 [00:05<00:00,  1.09it/s]\n",
      "Gene:\tSfpq-206: 100%|██████████| 11/11 [00:07<00:00,  1.46it/s]\n",
      "Gene:\tLars2-205: 100%|██████████| 4/4 [00:02<00:00,  1.54it/s]\n",
      "Gene:\tJakmip1-203: 100%|██████████| 16/16 [00:11<00:00,  1.36it/s]\n",
      "Gene:\tAdgrb1-204: 100%|██████████| 11/11 [00:06<00:00,  1.65it/s]\n",
      "Gene:\tPtprs-215: 100%|██████████| 9/9 [00:06<00:00,  1.35it/s]\n",
      "Gene:\tTmem145-201: 100%|██████████| 4/4 [00:02<00:00,  1.50it/s]\n",
      "Gene:\tGm17571-201: 100%|██████████| 2/2 [00:01<00:00,  1.60it/s]\n",
      "Gene:\tPtprn-202: 100%|██████████| 5/5 [00:02<00:00,  1.93it/s]\n",
      "Gene:\tAdgrb1-208: 100%|██████████| 8/8 [00:07<00:00,  1.12it/s]\n",
      "Gene:\tZmynd8-207: 100%|██████████| 18/18 [00:12<00:00,  1.43it/s]\n",
      "Gene:\tAtxn2l-202: 100%|██████████| 6/6 [00:05<00:00,  1.16it/s]\n",
      "Gene:\tGm26945-201: 100%|██████████| 34/34 [00:25<00:00,  1.32it/s]\n",
      "Gene:\tRasgrf1-204: 100%|██████████| 4/4 [00:02<00:00,  1.47it/s]\n",
      "Gene:\tCelf5-207: 100%|██████████| 9/9 [00:06<00:00,  1.31it/s]\n",
      "Gene:\tRpl36-201: 100%|██████████| 1/1 [00:00<00:00,  1.86it/s]\n",
      "Gene:\tLeng8-201: 100%|██████████| 11/11 [00:09<00:00,  1.17it/s]\n",
      "Gene:\tNdufb2-203: 100%|██████████| 4/4 [00:02<00:00,  1.85it/s]\n",
      "Gene:\tCamk2a-203: 100%|██████████| 9/9 [00:06<00:00,  1.47it/s]\n",
      "Gene:\tAdgrb1-212: 100%|██████████| 8/8 [00:04<00:00,  1.90it/s]\n",
      "Gene:\tKctd16-201: 100%|██████████| 12/12 [00:06<00:00,  1.87it/s]\n",
      "Gene:\tNdrg4-202: 100%|██████████| 11/11 [00:10<00:00,  1.04it/s]\n",
      "Gene:\tNdrg4-205: 100%|██████████| 10/10 [00:05<00:00,  1.93it/s]\n",
      "Gene:\tNdrg4-211: 100%|██████████| 9/9 [00:07<00:00,  1.25it/s]\n",
      "Gene:\t2900045O20Rik-201: 100%|██████████| 1/1 [00:01<00:00,  1.02s/it]\n",
      "Gene:\tGm42418-201: 100%|██████████| 1/1 [00:00<?, ?it/s]\n",
      "Gene:\tGm26917-202: 100%|██████████| 1/1 [00:00<00:00, 1002.46it/s]\n",
      "Gene:\tNdrg4-206: 100%|██████████| 10/10 [00:05<00:00,  1.95it/s]\n",
      "Gene:\tGm20594-201: 100%|██████████| 1/1 [00:00<00:00,  1.69it/s]\n",
      "Gene:\tAY036118-201: 100%|██████████| 1/1 [00:00<?, ?it/s]\n",
      "Gene:\tLars2-209: 100%|██████████| 9/9 [00:07<00:00,  1.25it/s]\n"
     ]
    }
   ],
   "source": [
    "from lib.database_interaction import ensembl_id_to_seqs\n",
    "import time\n",
    "\n",
    "skip = 0\n",
    "trial = 0\n",
    "\n",
    "tmp_isoform_list = isoform_list[skip:]\n",
    "tmp_id_list = id_list[len(sequences_of_all)+skip:]\n",
    "\n",
    "for i in range(len(tmp_isoform_list)):\n",
    "    isoform = tmp_isoform_list[i]\n",
    "    id = tmp_id_list[i]\n",
    "    sequences_of_all[f'{id}_{isoform}'] = dict()\n",
    "    sequences = ensembl_id_to_seqs(gene=isoform, gene_id=id.split('.')[0], seq_type='cds')\n",
    "    for desc, sequence in sequences.items():\n",
    "        sequences_of_all[f'{id}_{isoform}'][desc] = sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Gene:\tGm16024-201: 100%|██████████| 1/1 [00:00<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "from lib.database_interaction import ensembl_name_to_seqs\n",
    "import time\n",
    "\n",
    "skip = 0\n",
    "trial = 0\n",
    "\n",
    "for i in range(1):\n",
    "    if trial > 3:\n",
    "        skip += 1\n",
    "        trial = 0\n",
    "    try:\n",
    "        tmp_isoform_list = isoform_list[skip:]\n",
    "        tmp_id_list = id_list[len(sequences_of_all)+skip:]\n",
    "        for i in range(len(tmp_isoform_list)):\n",
    "            isoform = tmp_isoform_list[i]\n",
    "            sequences = ensembl_name_to_seqs(gene=isoform, species='mouse', seq_type='cds')\n",
    "            for desc, sequence in sequences.items():\n",
    "                sequences_of_all[desc] = sequence\n",
    "        skip += 1\n",
    "    except: \n",
    "        trial += 1\n",
    "        time.sleep(5)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Binding site Searcher\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from lib.search_binding import step_by_step, find_max_min_difference_fixed_length_subsequence, seq_minus\n",
    "\n",
    "# Initiation of array\n",
    "binding_site_FOIs = [\n",
    "    \"accession\",\n",
    "    \"gene_name\",\n",
    "    \"mol_type\",\n",
    "    \"organism\",\n",
    "    \"pos_on_seq\",\n",
    "    \"binding\",\n",
    "    \"Tm_l\",\n",
    "    \"Tm_r\",\n",
    "    \"wanted\",\n",
    "]\n",
    "align_FOIs = [\"align_num\", \"align_accession\", \"align_descrip\", \"plus/minus\"]\n",
    "FOI = pd.DataFrame(columns=binding_site_FOIs + align_FOIs)\n",
    "\n",
    "# Search binding sites on mRNA sequence\n",
    "file_out_dir = pre_binding_dir\n",
    "try:\n",
    "    os.mkdir(file_out_dir)\n",
    "except:\n",
    "    pass\n",
    "\n",
    "pre_binding_num = {}\n",
    "\n",
    "# initialization of file\n",
    "with open(os.path.join(file_out_dir, total_pre_binding_file_name), \"w\") as handle:\n",
    "    handle.write(\"\")\n",
    "\n",
    "for desc, seq in sequences_of_all.items():\n",
    "    id, gene_name, mol_type = desc.split('|')\n",
    "    minus_seq = seq_minus(seq)\n",
    "    \n",
    "    Tm_l, Tm_r, selected_substrings, pos_on_seq = step_by_step(\n",
    "        minus_seq,\n",
    "        BDS_len=40,\n",
    "        BDS_num=50,\n",
    "        min_gap=1,\n",
    "        better_gap=40,\n",
    "        gene=gene_name,\n",
    "        G_min=0.25,\n",
    "        G_max=0.7,\n",
    "        G_consecutive=5,\n",
    "        Tm_low=50,\n",
    "        Tm_high=65,\n",
    "    )\n",
    "    \n",
    "    record_list = []\n",
    "    for i, pre_binding_tmp in enumerate(selected_substrings):\n",
    "        record_list.append(\n",
    "            SeqRecord(\n",
    "                Seq(pre_binding_tmp),\n",
    "                id=\"pre_binding\" + str(i),\n",
    "                description=\"|\".join([id, gene_name, organism, mol_type]),\n",
    "            )\n",
    "        )\n",
    "\n",
    "    # add information about binding sites to FOI\n",
    "    add = pd.DataFrame(\n",
    "        {\n",
    "            \"accession\": [id] * len(selected_substrings),\n",
    "            \"gene_name\": [gene_name] * len(selected_substrings),\n",
    "            \"mol_type\": [mol_type] * len(selected_substrings),\n",
    "            \"organism\": [organism] * len(selected_substrings),\n",
    "            \"binding\": selected_substrings,\n",
    "            \"Tm_l\": Tm_l,\n",
    "            \"Tm_r\": Tm_r,\n",
    "            \"pos_on_seq\": pos_on_seq,\n",
    "        }\n",
    "    )\n",
    "    FOI = pd.concat([FOI, add], ignore_index=True)\n",
    "\n",
    "    file_out = os.path.join(file_out_dir, gene_name + pre_binding_file_suffix)\n",
    "    \n",
    "    # write pre_binding to files\n",
    "    with open(file_out, \"w\") as f:\n",
    "        for new_record in record_list:\n",
    "            SeqIO.write(new_record, f, \"fasta\")\n",
    "    with open(file_out_dir + total_pre_binding_file_name, \"a\") as handle:\n",
    "        for new_record in record_list:\n",
    "            SeqIO.write(new_record, handle, \"fasta\")\n",
    "\n",
    "    # record the num of pre_binding for each gene\n",
    "    pre_binding_num[f\"{id}_{gene_name}\"] = len(selected_substrings)\n",
    "\n",
    "with open(os.path.join(output, pre_binding_num_file), \"w\") as f:\n",
    "    json.dump(pre_binding_num, f)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blast and extract blast results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open(file_out_dir + total_pre_binding_file_name, \"r\") as f:\n",
    "#     fasta_string = f.read()\n",
    "# txid = [2697049]  # organism\n",
    "\n",
    "# # Submit BLAST search and get handle object\n",
    "# handle = NCBIWWW.qblast(\n",
    "#     program=\"blastn\",\n",
    "#     megablast=\"yes\",\n",
    "#     database=\"refseq_rna\",\n",
    "#     sequence=fasta_string,\n",
    "#     url_base=\"https://blast.ncbi.nlm.nih.gov/Blast.cgi\",\n",
    "#     format_object=\"Alignment\",\n",
    "#     format_type=\"Xml\",\n",
    "# )\n",
    "\n",
    "# # read handle object and save to a file\n",
    "# with open(os.path.join(os.path.join(output, blast_results_file)), \"w\") as f:\n",
    "#     f.write(handle.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract interested information from blast_results\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "\n",
    "align_num = []\n",
    "# read the id/plus-minus part/align_num\n",
    "with open(os.path.join(output, blast_results_file), \"r\") as blast_output:\n",
    "    blast_records = NCBIXML.parse(blast_output)\n",
    "    loca = 0\n",
    "    for blast_record in blast_records:\n",
    "        align_accession = []\n",
    "        align_descrip_list = []\n",
    "        # get align num of each binding site\n",
    "        length = len(blast_record.alignments)\n",
    "        align_num.append(length)\n",
    "        for i in range(length):\n",
    "            descrip = blast_record.descriptions[i].title.split(\"|\")\n",
    "            # get accession and descrip of each align seq\n",
    "            align_accession.append(descrip[3])\n",
    "            align_descrip_list.append(descrip[-1])\n",
    "        FOI.loc[loca, \"align_accession\"] = \"|\".join(str(_) for _ in align_accession)\n",
    "\n",
    "        # add align_descrip to df\n",
    "        FOI.loc[loca, \"align_descrip\"] = \"|\".join(str(_) for _ in align_descrip_list)\n",
    "\n",
    "        # get plus/minus of each align seq\n",
    "        p_m = [blast_record.alignments[_].hsps[0].frame[1] for _ in range(length)]\n",
    "\n",
    "        # add plus/minus to df\n",
    "        try:\n",
    "            FOI.loc[loca, \"plus/minus\"] = \",\".join([str(_) for _ in p_m])\n",
    "        except:\n",
    "            FOI.loc[loca, \"plus/minus\"] = \"NAN\"\n",
    "\n",
    "        loca += 1\n",
    "\n",
    "FOI[\"align_num\"] = align_num"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select wanted binding site\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "FOI[\"wanted\"] = [True] * len(FOI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sieve for the suitable binding site\n",
    "gene_name_list = [_.upper() for _ in gene_list]\n",
    "gene_name_list_out = [i for i in gene_name_list]\n",
    "for i in range(len(FOI)):\n",
    "    # check gene_name\n",
    "    gene_name = FOI.loc[i, \"gene_name\"]\n",
    "    if gene_name.split('-')[0].upper() not in gene_name_list:\n",
    "        FOI.loc[i, \"wanted\"] = False\n",
    "    else:\n",
    "        try:\n",
    "            gene_name_list_out.remove(gene_name)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # check DNA or mRNA type\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        if FOI.loc[i, \"mol_type\"] != \"protein_coding\":\n",
    "            FOI.loc[i, \"wanted\"] = False\n",
    "            print(FOI.loc[i, \"mol_type\"])\n",
    "\n",
    "    # check gene_organism name\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        spe_ori, gene_ori = FOI.loc[i, \"organism\"], FOI.loc[i, \"gene_name\"].split('-')[0]\n",
    "        descrip = FOI.loc[i, \"align_descrip\"].split(\"|\")\n",
    "        for des in descrip:\n",
    "            if gene_ori not in des and spe_ori in des:\n",
    "                FOI.loc[i, \"wanted\"] = False\n",
    "                break\n",
    "\n",
    "    # check plus/minus\n",
    "    if FOI.loc[i, \"wanted\"] == True:\n",
    "        if pd.isnull(FOI.loc[i, \"plus/minus\"]):\n",
    "            FOI.loc[i, \"wanted\"] = False\n",
    "        else:\n",
    "            pm_list = FOI.loc[i, \"plus/minus\"].split(\",\")\n",
    "            if \"-1\" not in pm_list:\n",
    "                FOI.loc[i, \"wanted\"] = False\n",
    "\n",
    "# write the whole information of interest to a excel file in tmp dir\n",
    "FOI.to_excel(os.path.join(output, \"probes_sieve.xlsx\"))\n",
    "\n",
    "out_tmp = FOI[FOI[\"wanted\"] == True]\n",
    "output_df = pd.DataFrame()\n",
    "for gene in out_tmp.gene_name.unique():\n",
    "    pos_of_True = list(out_tmp[out_tmp.gene_name == gene][\"pos_on_seq\"])\n",
    "    best_pos = find_max_min_difference_fixed_length_subsequence(\n",
    "        pos_of_True,\n",
    "        length=3,\n",
    "        min_gap=40,\n",
    "        better_gap=80,\n",
    "        gene=gene,\n",
    "    )\n",
    "    out_subset = out_tmp[out_tmp.gene_name == gene]\n",
    "    out_subset = out_subset[out_subset[\"pos_on_seq\"].isin(best_pos)]\n",
    "    output_df = pd.concat([output_df, out_subset])\n",
    "\n",
    "# write the output to a xlsx file\n",
    "output_df.to_excel(os.path.join(output, \"probes_wanted.xlsx\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
