{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Probe Designer\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basci env\n",
    "import os\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import time\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "\n",
    "# data process of file from ncbi\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "# dir\n",
    "DATASET_DIR = Path('/mnt/f/spatial_data/probe')\n",
    "RUNID = 'Example_dataset'\n",
    "workdir = DATASET_DIR / RUNID\n",
    "os.makedirs(workdir, exist_ok=True)\n",
    "organism = 'human'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create results dir\n",
    "current_time = time.localtime()\n",
    "formatted_time = time.strftime(\"%Y%m%d_%H%M%S\", current_time)\n",
    "output = os.path.join(workdir, 'results', formatted_time+'_ensembl')\n",
    "bds_candidate_dir = os.path.join(output, \"bds_candidate\")\n",
    "os.makedirs(output, exist_ok=True)\n",
    "os.makedirs(bds_candidate_dir, exist_ok=True)\n",
    "\n",
    "# file name variables\n",
    "bds_candidate_file_suffix = \"_bds_candidate.fasta\"\n",
    "combined_bds_candidates_file = \"total_bds_candidate.fasta\"\n",
    "combined_bds_candidates_blast_file = 'total_bds_candidate_blast.fasta'\n",
    "bds_candidate_num_file = \"bds_candidate_num.json\"\n",
    "blast_results_file = \"blast_results.xml\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_info = pd.read_excel(os.path.join(workdir, \"marker_gene_list.xlsx\"))\n",
    "gene_info['gene'] = gene_info['gene name']\n",
    "if organism == 'mouse': gene_info['gene'] = gene_info['gene'].str.capitalize()\n",
    "elif organism == 'human': gene_info['gene'] = gene_info['gene'].str.upper()\n",
    "gene_list = [_.strip() for _ in gene_info['gene'].unique() if _!=0]\n",
    "print(len(gene_list))\n",
    "gene_info.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get seq from ensembl dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from lib.database_interaction import ensembl_id_to_seqs\n",
    "# import time\n",
    "\n",
    "# skip = 0\n",
    "# trial = 0\n",
    "\n",
    "# tmp_isoform_list = isoform_list[skip:]\n",
    "# tmp_id_list = id_list[len(sequences_of_all)+skip:]\n",
    "\n",
    "# for i in range(len(tmp_isoform_list)):\n",
    "#     isoform = tmp_isoform_list[i]\n",
    "#     id = tmp_id_list[i]\n",
    "#     sequences_of_all[f'{id}_{isoform}'] = dict()\n",
    "#     sequences = ensembl_id_to_seqs(gene=isoform, gene_id=id.split('.')[0], seq_type='cds')\n",
    "#     for desc, sequence in sequences.items():\n",
    "#         sequences_of_all[f'{id}_{isoform}'][desc] = sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from lib.database_interaction import ensembl_name_to_seqs\n",
    "import time\n",
    "\n",
    "\n",
    "max_trial = 3\n",
    "sequences_of_all = dict()\n",
    "error_messages = {gene: [] for gene in gene_list}\n",
    "\n",
    "with tqdm(total=len(gene_list), desc=\"Retriving_sequences\", position=0) as pbar_total:\n",
    "    for gene in gene_list:\n",
    "        sequences_of_all[gene] = {}\n",
    "        trial_success = False\n",
    "        \n",
    "        # Reset the trial progress bar for each gene\n",
    "        for trial in range(1, max_trial+1):  # Retrying up to 3 times\n",
    "            try:\n",
    "                # Attempt to retrieve sequences\n",
    "                sequences_of_all[gene] = ensembl_name_to_seqs(gene=gene, species=organism, seq_type='cds', tqdm_args={'position': 1,'leave': False})\n",
    "                trial_success = True\n",
    "                break\n",
    "            except Exception as e:\n",
    "                time.sleep(1)\n",
    "\n",
    "        if not trial_success:\n",
    "            error_messages[gene].append(f\"Failed to retrieve sequences for {gene} after {max_trial} attempts.\")\n",
    "\n",
    "        pbar_total.update(1)  # Update the main progress bar after each gene\n",
    "\n",
    "for gene, messages in error_messages.items():\n",
    "    for message in messages:\n",
    "        print(message)\n",
    "\n",
    "with open(os.path.join(output, 'sequence_of_all.json'), 'w') as file: json.dump(sequences_of_all, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(output, 'sequence_of_all.json'), 'r') as file: sequences_of_all = json.load(file)\n",
    "\n",
    "isoforms = {}\n",
    "for gene, isoforms_tmp in sequences_of_all.items():\n",
    "    shortest_isoform = None\n",
    "    min_length = 10**6\n",
    "    max_length = 0\n",
    "    for isoform in isoforms_tmp:\n",
    "        try: isoform_name = isoform['external_name']\n",
    "        except: continue\n",
    "        if len(isoform['seq']) < min_length and gene.upper() in isoform_name.upper():\n",
    "            min_length = len(isoform['seq'])\n",
    "            shortest_isoform = isoform\n",
    "    if shortest_isoform:\n",
    "        isoforms[gene] = shortest_isoform\n",
    "with open(os.path.join(output, 'shortest_isoforms.json'), 'w') as file: json.dump(isoforms, file)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Binding site Searcher\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# set lib auto reload in jupyter notebook\n",
    "from lib.search_binding import position_search, optimize_subsequence, seq_minus\n",
    "\n",
    "# Initiation of array\n",
    "binding_site_entry = [\n",
    "    \"accession\", \"gene_name\", \"mol_type\", \"organism\",\n",
    "    \"pos\", \"bds\", \"Tm\", \"Tm_l\", \"Tm_r\", \"mfe\", \"wanted\"]\n",
    "alignment_entry = [\"align_num\", \"align_accession\", \"align_descrip\", \"plus/minus\"]\n",
    "BDS_INFO = pd.DataFrame(columns=binding_site_entry+alignment_entry)\n",
    "\n",
    "# Search binding sites on mRNA sequence\n",
    "pre_binding_num = {}\n",
    "\n",
    "# initialization of file\n",
    "with open(os.path.join(output, combined_bds_candidates_file), \"w\") as handle: handle.write(\"\")\n",
    "with open(os.path.join(output, combined_bds_candidates_blast_file), \"w\") as f: f.write(\"\")\n",
    "\n",
    "for desc, info in tqdm(isoforms.items(), desc=\"Searching_binding_sites\", position=0):\n",
    "    seq = info['seq']\n",
    "    if 'N' in seq: seq = seq.replace('N', '')\n",
    "    try: gene_name= info['external_name']\n",
    "    except: gene_name = desc\n",
    "\n",
    "    id = info['id']\n",
    "    mol_type = info['biotype']\n",
    "    minus_seq = seq_minus(seq)\n",
    "    \n",
    "    pos_info = position_search(\n",
    "        minus_seq, gene=gene_name,\n",
    "        BDS_len=40, BDS_num=50, min_gap=0, better_gap=40,\n",
    "        G_min=0.25, G_max=0.7, G_consecutive=5, Tm_low=48, Tm_high=60, \n",
    "        verbose_pos=1, leave=False, warn=False)\n",
    "    \n",
    "\n",
    "    record_list = []\n",
    "    for i, pre_binding_tmp in enumerate([_['bds'] for _ in pos_info]):\n",
    "        record_list.append(\n",
    "            SeqRecord(\n",
    "                Seq(pre_binding_tmp),\n",
    "                id=\"bds_candidate\" + str(i),\n",
    "                description=\"|\".join([id, gene_name, organism, mol_type])))\n",
    "\n",
    "    # add information about binding sites to FOI\n",
    "    add = pd.DataFrame(pos_info)\n",
    "    add['accession'] = id\n",
    "    add['gene_name'] = gene_name\n",
    "    add['mol_type'] = mol_type\n",
    "    add['organism'] = organism\n",
    "    BDS_INFO = pd.concat([BDS_INFO, add], ignore_index=True)\n",
    "\n",
    "    file_out = os.path.join(bds_candidate_dir, gene_name + bds_candidate_file_suffix)\n",
    "    \n",
    "    # write pre_binding to files\n",
    "    with open(file_out, \"w\") as f:\n",
    "        for new_record in record_list: SeqIO.write(new_record, f, \"fasta\")\n",
    "    with open(os.path.join(output, combined_bds_candidates_file), \"a\") as handle:\n",
    "        for new_record in record_list: SeqIO.write(new_record, handle, \"fasta\")\n",
    "    with open(os.path.join(output, combined_bds_candidates_blast_file), \"a\") as handle:\n",
    "        for new_record in record_list: \n",
    "            seq = str(new_record.seq)\n",
    "            seq = seq[len(seq)//2-16:len(seq)//2+16]\n",
    "            new_record = SeqRecord(Seq(seq), id=new_record.id, description=new_record.description)\n",
    "            SeqIO.write(new_record, handle, \"fasta\")\n",
    "\n",
    "    # record the num of pre_binding for each gene\n",
    "    pre_binding_num[f\"{id}_{gene_name}\"] = len(pos_info)\n",
    "\n",
    "with open(os.path.join(output, bds_candidate_num_file), \"w\") as f:\n",
    "    json.dump(pre_binding_num, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blast and extract blast results\n",
    "- NCBIXML: https://homolog.us/Biopython/Bio.Blast.NCBIXML.html#read/0\n",
    "- BlastRecord: https://biopython.org/docs/1.75/api/Bio.Blast.Record.html\n",
    "- XMLReader: https://codebeautify.org/xmlviewer#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open(file_out_dir + total_pre_binding_file_name, \"r\") as f: fasta_string = f.read()\n",
    "# txid = [2697049]  # organism\n",
    "\n",
    "# # Submit BLAST search and get handle object\n",
    "# handle = NCBIWWW.qblast(\n",
    "#     program=\"blastn\",\n",
    "#     megablast=\"yes\",\n",
    "#     database=\"refseq_rna\",\n",
    "#     sequence=fasta_string,\n",
    "#     url_base=\"https://blast.ncbi.nlm.nih.gov/Blast.cgi\",\n",
    "#     format_object=\"Alignment\",\n",
    "#     format_type=\"Xml\",\n",
    "# )\n",
    "\n",
    "# # read handle object and save to a file\n",
    "# with open(os.path.join(os.path.join(output, blast_results_file)), \"w\") as f: f.write(handle.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract interested information from blast_results\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "\n",
    "align_num = []\n",
    "# read the id/plus-minus part/align_num\n",
    "with open(os.path.join(output, blast_results_file), \"r\") as blast_output:\n",
    "    blast_records = NCBIXML.parse(blast_output)\n",
    "    loca = 0\n",
    "    for blast_record in blast_records:\n",
    "        align_accession = []\n",
    "        align_descrip_list = []\n",
    "        # get align num of each binding site\n",
    "        length = len(blast_record.alignments)\n",
    "        align_num.append(length)\n",
    "        for i in range(length):\n",
    "            descrip = blast_record.descriptions[i].title.split(\"|\")\n",
    "            # get accession and descrip of each align seq\n",
    "            align_accession.append(descrip[3])\n",
    "            align_descrip_list.append(descrip[-1])\n",
    "        BDS_INFO.loc[loca, \"align_accession\"] = \"|\".join(str(_) for _ in align_accession)\n",
    "        # add align_descrip to df\n",
    "        BDS_INFO.loc[loca, \"align_descrip\"] = \"|\".join(str(_) for _ in align_descrip_list)\n",
    "        # get plus/minus of each align seq\n",
    "        p_m = [blast_record.alignments[_].hsps[0].frame[1] for _ in range(length)]\n",
    "        # add plus/minus to df\n",
    "        try: BDS_INFO.loc[loca, \"plus/minus\"] = \",\".join([str(_) for _ in p_m])\n",
    "        except: BDS_INFO.loc[loca, \"plus/minus\"] = pd.NA\n",
    "        loca += 1\n",
    "BDS_INFO[\"align_num\"] = align_num"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select wanted binding site\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def adjust_gene_name(gene_name, gene_list):\n",
    "    gene_list = [x.upper() for x in gene_list]\n",
    "    match = re.search(r'(.+)-(\\d+)$', gene_name)\n",
    "    if match:\n",
    "        base_gene_name = match.group(1)\n",
    "        if base_gene_name.upper() in gene_list or gene_name.upper() in gene_list: return base_gene_name\n",
    "        else: return gene_name\n",
    "    else: return gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "BDS_INFO[\"wanted\"] = [True] * len(BDS_INFO)\n",
    "verbose = True\n",
    "# select by specifity\n",
    "gene_name_list = [_.upper() for _ in gene_list]\n",
    "gene_name_list_out = [i for i in gene_name_list]\n",
    "for i in range(len(BDS_INFO)):\n",
    "    # check gene_name\n",
    "    gene_name = adjust_gene_name(BDS_INFO.loc[i, \"gene_name\"], gene_name_list)\n",
    "    spe_ori= BDS_INFO.loc[i, \"organism\"]\n",
    "    if gene_name.upper() not in gene_name_list: \n",
    "        BDS_INFO.loc[i, \"wanted\"] = False\n",
    "        if verbose: print(f\"{gene_name} not in gene list.\")\n",
    "    else:\n",
    "        try: gene_name_list_out.remove(gene_name)\n",
    "        except: pass\n",
    "\n",
    "    # check DNA or mRNA type\n",
    "    if BDS_INFO.loc[i, \"wanted\"] == True:\n",
    "        if BDS_INFO.loc[i, \"mol_type\"] != \"protein_coding\":\n",
    "            # BDS_INFO.loc[i, \"wanted\"] = False\n",
    "            if verbose: print(\"{} is {}.\".format(gene_name, BDS_INFO.loc[i, \"mol_type\"]))\n",
    "\n",
    "    # check gene_organism name\n",
    "    if BDS_INFO.loc[i, \"wanted\"] == True:\n",
    "        descrip = BDS_INFO.loc[i, \"align_descrip\"]\n",
    "        if pd.isnull(descrip):\n",
    "            BDS_INFO.loc[i, \"wanted\"] = False\n",
    "            if verbose: print(f\"{gene_name} not found in BLAST.\")\n",
    "        else:\n",
    "            descrip = descrip.split(\"|\")\n",
    "            for des in descrip:\n",
    "                if gene_name not in des and spe_ori in des:\n",
    "                    BDS_INFO.loc[i, \"wanted\"] = False\n",
    "                    if verbose: print(f\"{gene_name} not specific.\")\n",
    "                    break\n",
    "\n",
    "    # check plus/minus\n",
    "    if BDS_INFO.loc[i, \"wanted\"] == True:\n",
    "        pm_list = BDS_INFO.loc[i, \"plus/minus\"].split(\",\")\n",
    "        if \"-1\" not in pm_list:\n",
    "            BDS_INFO.loc[i, \"wanted\"] = False\n",
    "            if verbose: print(f\"{gene_name} not plus/minus.\")\n",
    "\n",
    "# write the whole information of interest to a excel file in tmp dir\n",
    "BDS_INFO.to_excel(os.path.join(output, \"probes_candidates.xlsx\"))\n",
    "\n",
    "out_tmp = BDS_INFO[BDS_INFO[\"wanted\"] == True]\n",
    "output_df = pd.DataFrame()\n",
    "for gene in out_tmp.gene_name.unique():\n",
    "    pos_wanted = list(out_tmp[out_tmp.gene_name == gene][\"pos\"])\n",
    "    pos_best = optimize_subsequence(pos_wanted, length=3, min_gap=40, better_gap=80, gene=gene)\n",
    "    pos_output = out_tmp[out_tmp.gene_name == gene]\n",
    "    pos_output = pos_output[pos_output[\"pos\"].isin(pos_best)]\n",
    "    output_df = pd.concat([output_df, pos_output])\n",
    "\n",
    "# write the output to a xlsx file\n",
    "output_df.to_excel(os.path.join(output, \"probes_wanted.xlsx\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
