{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Probe Designer\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basci env\n",
    "import os\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import time\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "\n",
    "# data process of file from ncbi\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "# dir\n",
    "DATASET_DIR = Path('F:\\spatial_data\\dataset_probe')\n",
    "RUNID = '20240808_Z.Y_Mouse'\n",
    "workdir = DATASET_DIR / RUNID\n",
    "os.makedirs(workdir, exist_ok=True)\n",
    "organism = 'mouse'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create results dir\n",
    "current_time = time.localtime()\n",
    "formatted_time = time.strftime(\"%Y%m%d_%H%M%S\", current_time)\n",
    "output = os.path.join(workdir, 'results', formatted_time+'_ensembl')\n",
    "bds_candidate_dir = os.path.join(output, \"bds_candidate\")\n",
    "os.makedirs(output, exist_ok=True)\n",
    "os.makedirs(bds_candidate_dir, exist_ok=True)\n",
    "\n",
    "# file name variables\n",
    "bds_candidate_file_suffix = \"_bds_candidate.fasta\"\n",
    "combined_bds_candidates_file = \"total_bds_candidate.fasta\"\n",
    "combined_bds_candidates_blast_file = 'total_bds_candidate_blast.fasta'\n",
    "bds_candidate_num_file = \"bds_candidate_num.json\"\n",
    "blast_results_file = \"blast_results.xml\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Remark</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tnni3</td>\n",
       "      <td>心肌细胞</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Actn2</td>\n",
       "      <td>心肌细胞</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Myl2</td>\n",
       "      <td>心肌细胞</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Nppa</td>\n",
       "      <td>心肌细胞</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ankrd1</td>\n",
       "      <td>心肌细胞</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene Remark\n",
       "0   Tnni3   心肌细胞\n",
       "1   Actn2   心肌细胞\n",
       "2    Myl2   心肌细胞\n",
       "3    Nppa   心肌细胞\n",
       "4  Ankrd1   心肌细胞"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info = pd.read_excel(os.path.join(workdir, \"Marker_genes.xlsx\"))\n",
    "if organism == 'mouse': gene_info['gene'] = gene_info['gene'].str.capitalize()\n",
    "elif organism == 'human': gene_info['gene'] = gene_info['gene'].str.upper()\n",
    "gene_list = [_.strip() for _ in gene_info['gene'].unique() if _!=0]\n",
    "print(len(gene_list))\n",
    "gene_info.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get seq from ensembl dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from lib.database_interaction import ensembl_id_to_seqs\n",
    "# import time\n",
    "\n",
    "# skip = 0\n",
    "# trial = 0\n",
    "\n",
    "# tmp_isoform_list = isoform_list[skip:]\n",
    "# tmp_id_list = id_list[len(sequences_of_all)+skip:]\n",
    "\n",
    "# for i in range(len(tmp_isoform_list)):\n",
    "#     isoform = tmp_isoform_list[i]\n",
    "#     id = tmp_id_list[i]\n",
    "#     sequences_of_all[f'{id}_{isoform}'] = dict()\n",
    "#     sequences = ensembl_id_to_seqs(gene=isoform, gene_id=id.split('.')[0], seq_type='cds')\n",
    "#     for desc, sequence in sequences.items():\n",
    "#         sequences_of_all[f'{id}_{isoform}'][desc] = sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retriving_sequences: 100%|██████████| 29/29 [01:54<00:00,  3.96s/it]\n"
     ]
    }
   ],
   "source": [
    "from lib.database_interaction import ensembl_name_to_seqs\n",
    "import time\n",
    "\n",
    "\n",
    "max_trial = 3\n",
    "sequences_of_all = dict()\n",
    "error_messages = {gene: [] for gene in gene_list}\n",
    "\n",
    "with tqdm(total=len(gene_list), desc=\"Retriving_sequences\", position=0) as pbar_total:\n",
    "    for gene in gene_list:\n",
    "        sequences_of_all[gene] = {}\n",
    "        trial_success = False\n",
    "        \n",
    "        # Reset the trial progress bar for each gene\n",
    "        for trial in range(1, max_trial+1):  # Retrying up to 3 times\n",
    "            try:\n",
    "                # Attempt to retrieve sequences\n",
    "                sequences_of_all[gene] = ensembl_name_to_seqs(gene=gene, species=organism, seq_type='cds', tqdm_args={'position': 1,'leave': False})\n",
    "                trial_success = True\n",
    "                break\n",
    "            except Exception as e:\n",
    "                time.sleep(1)\n",
    "\n",
    "        if not trial_success:\n",
    "            error_messages[gene].append(f\"Failed to retrieve sequences for {gene} after {max_trial} attempts.\")\n",
    "\n",
    "        pbar_total.update(1)  # Update the main progress bar after each gene\n",
    "\n",
    "for gene, messages in error_messages.items():\n",
    "    for message in messages:\n",
    "        print(message)\n",
    "\n",
    "with open(os.path.join(output, 'sequence_of_all.json'), 'w') as file: json.dump(sequences_of_all, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(output, 'sequence_of_all.json'), 'r') as file: sequences_of_all = json.load(file)\n",
    "\n",
    "isoforms = {}\n",
    "for gene, isoforms_tmp in sequences_of_all.items():\n",
    "    shortest_isoform = None\n",
    "    min_length = 10**6\n",
    "    max_length = 0\n",
    "    for isoform in isoforms_tmp:\n",
    "        try: isoform_name = isoform['external_name']\n",
    "        except: continue\n",
    "        if len(isoform['seq']) < min_length and gene.upper() in isoform_name.upper():\n",
    "            min_length = len(isoform['seq'])\n",
    "            shortest_isoform = isoform\n",
    "    if shortest_isoform:\n",
    "        isoforms[gene] = shortest_isoform\n",
    "with open(os.path.join(output, 'shortest_isoforms.json'), 'w') as file: json.dump(isoforms, file)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Binding site Searcher\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Searching_binding_sites:   0%|          | 0/29 [00:00<?, ?it/s]C:\\Users\\Mingchuan\\AppData\\Local\\Temp\\ipykernel_34932\\2255783389.py:57: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  BDS_INFO = pd.concat([BDS_INFO, add], ignore_index=True)\n",
      "Searching_binding_sites: 100%|██████████| 29/29 [00:00<00:00, 29.74it/s]\n"
     ]
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# set lib auto reload in jupyter notebook\n",
    "from lib.search_binding import position_search, optimize_subsequence, seq_minus\n",
    "\n",
    "# Initiation of array\n",
    "binding_site_entry = [\"accession\", \"gene_name\", \"mol_type\", \"organism\", \n",
    "                     \"pos_on_seq\", \"binding\", \"Tm_l\", \"Tm_r\", \"wanted\",]\n",
    "alignment_entry = [\"align_num\", \"align_accession\", \"align_descrip\", \"plus/minus\"]\n",
    "BDS_INFO = pd.DataFrame(columns=binding_site_entry+alignment_entry)\n",
    "\n",
    "# Search binding sites on mRNA sequence\n",
    "pre_binding_num = {}\n",
    "\n",
    "# initialization of file\n",
    "with open(os.path.join(output, combined_bds_candidates_file), \"w\") as handle: handle.write(\"\")\n",
    "with open(os.path.join(output, combined_bds_candidates_blast_file), \"w\") as f: f.write(\"\")\n",
    "\n",
    "for desc, info in tqdm(isoforms.items(), desc=\"Searching_binding_sites\", position=0):\n",
    "    seq = info['seq']\n",
    "    if 'N' in seq: seq = seq.replace('N', '')\n",
    "    try: gene_name= info['external_name']\n",
    "    except: gene_name = desc\n",
    "\n",
    "    id = info['id']\n",
    "    mol_type = info['biotype']\n",
    "    minus_seq = seq_minus(seq)\n",
    "    \n",
    "    pos_info = position_search(\n",
    "        minus_seq, gene=gene_name,\n",
    "        BDS_len=40, BDS_num=50, min_gap=1, better_gap=40,\n",
    "        G_min=0.25, G_max=0.7, G_consecutive=5, Tm_low=50, Tm_high=65, \n",
    "        verbose_pos=1, leave=False, warn=False)\n",
    "    \n",
    "\n",
    "    record_list = []\n",
    "    for i, pre_binding_tmp in enumerate([_['seq'] for _ in pos_info]):\n",
    "        record_list.append(\n",
    "            SeqRecord(\n",
    "                Seq(pre_binding_tmp),\n",
    "                id=\"pre_binding\" + str(i),\n",
    "                description=\"|\".join([id, gene_name, organism, mol_type])))\n",
    "\n",
    "    # add information about binding sites to FOI\n",
    "    add = pd.DataFrame(pos_info)\n",
    "    add['accession'] = id\n",
    "    add['gene_name'] = gene_name\n",
    "    add['mol_type'] = mol_type\n",
    "    add['organism'] = organism\n",
    "    BDS_INFO = pd.concat([BDS_INFO, add], ignore_index=True)\n",
    "\n",
    "    file_out = os.path.join(bds_candidate_dir, gene_name + bds_candidate_file_suffix)\n",
    "    \n",
    "    # write pre_binding to files\n",
    "    with open(file_out, \"w\") as f:\n",
    "        for new_record in record_list: SeqIO.write(new_record, f, \"fasta\")\n",
    "    with open(os.path.join(output, combined_bds_candidates_file), \"a\") as handle:\n",
    "        for new_record in record_list: SeqIO.write(new_record, handle, \"fasta\")\n",
    "    with open(os.path.join(output, combined_bds_candidates_blast_file), \"a\") as handle:\n",
    "        for new_record in record_list: \n",
    "            seq = str(new_record.seq)\n",
    "            seq = seq[len(seq)//2-16:len(seq)//2+16]\n",
    "            new_record = SeqRecord(Seq(seq), id=new_record.id, description=new_record.description)\n",
    "            SeqIO.write(new_record, handle, \"fasta\")\n",
    "\n",
    "    # record the num of pre_binding for each gene\n",
    "    pre_binding_num[f\"{id}_{gene_name}\"] = len(selected_substrings)\n",
    "\n",
    "with open(os.path.join(output, bds_candidate_num_file), \"w\") as f:\n",
    "    json.dump(pre_binding_num, f)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blast and extract blast results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open(file_out_dir + total_pre_binding_file_name, \"r\") as f: fasta_string = f.read()\n",
    "# txid = [2697049]  # organism\n",
    "\n",
    "# # Submit BLAST search and get handle object\n",
    "# handle = NCBIWWW.qblast(\n",
    "#     program=\"blastn\",\n",
    "#     megablast=\"yes\",\n",
    "#     database=\"refseq_rna\",\n",
    "#     sequence=fasta_string,\n",
    "#     url_base=\"https://blast.ncbi.nlm.nih.gov/Blast.cgi\",\n",
    "#     format_object=\"Alignment\",\n",
    "#     format_type=\"Xml\",\n",
    "# )\n",
    "\n",
    "# # read handle object and save to a file\n",
    "# with open(os.path.join(os.path.join(output, blast_results_file)), \"w\") as f:\n",
    "#     f.write(handle.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract interested information from blast_results\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "\n",
    "align_num = []\n",
    "# read the id/plus-minus part/align_num\n",
    "with open(os.path.join(output, blast_results_file), \"r\") as blast_output:\n",
    "    blast_records = NCBIXML.parse(blast_output)\n",
    "    loca = 0\n",
    "    for blast_record in blast_records:\n",
    "        align_accession = []\n",
    "        align_descrip_list = []\n",
    "        # get align num of each binding site\n",
    "        length = len(blast_record.alignments)\n",
    "        align_num.append(length)\n",
    "        for i in range(length):\n",
    "            descrip = blast_record.descriptions[i].title.split(\"|\")\n",
    "            # get accession and descrip of each align seq\n",
    "            align_accession.append(descrip[3])\n",
    "            align_descrip_list.append(descrip[-1])\n",
    "        BDS_INFO.loc[loca, \"align_accession\"] = \"|\".join(str(_) for _ in align_accession)\n",
    "        # add align_descrip to df\n",
    "        BDS_INFO.loc[loca, \"align_descrip\"] = \"|\".join(str(_) for _ in align_descrip_list)\n",
    "        # get plus/minus of each align seq\n",
    "        p_m = [blast_record.alignments[_].hsps[0].frame[1] for _ in range(length)]\n",
    "        # add plus/minus to df\n",
    "        try: BDS_INFO.loc[loca, \"plus/minus\"] = \",\".join([str(_) for _ in p_m])\n",
    "        except: BDS_INFO.loc[loca, \"plus/minus\"] = pd.NA\n",
    "        loca += 1\n",
    "BDS_INFO[\"align_num\"] = align_num"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select wanted binding site\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Ankrd1-203 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Cilp-202 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Comp-204 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Vwf-210 nonsense_mediated_decay\n",
      "Gene Tnni3-207: Not enough pos for 3 binding sites.\n",
      "Gene Tnni3-207: condition too harsh, loose to get better results\n",
      "[202, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 243, 244, 251, 255]\n",
      "Gene Cilp-202: Not enough pos for 3 binding sites.\n",
      "Gene Cilp-202: condition too harsh, loose to get better results\n",
      "[22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 106, 107, 108, 115, 116, 117, 118, 119, 120, 121, 122, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162]\n",
      "Gene Cthrc1-202: Not enough pos for 3 binding sites.\n",
      "Gene Cthrc1-202: condition too harsh, loose to get better results\n",
      "[39, 40, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 137, 141, 142, 143, 144, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 161, 162, 163, 164, 166, 167, 168, 169, 170, 171, 173]\n",
      "Gene Comp-204: Not enough pos for 3 binding sites.\n",
      "Gene Comp-204: condition too harsh, loose to get better results\n",
      "[49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 67, 68, 70, 71, 81, 82, 83, 84, 87, 88, 90, 91, 152, 155, 156, 157, 158]\n",
      "Gene Ccl2-201: Not enough pos for 3 binding sites.\n",
      "Gene Ccl2-201: condition too harsh, loose to get better results\n",
      "[255, 256, 259, 268, 279, 294, 304, 314, 315, 316, 324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334, 335, 336, 337, 338, 343, 344, 345, 346, 347, 348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 361, 362]\n",
      "Gene Cdh5-202: Not enough pos for 3 binding sites.\n",
      "Gene Cdh5-202: condition too harsh, loose to get better results\n",
      "[18, 19, 20, 24, 25, 26, 27, 28, 29, 30, 31, 32, 50, 51, 52]\n",
      "Gene Rgs5-202: Not enough pos for 3 binding sites.\n",
      "Gene Rgs5-202: condition too harsh, loose to get better results\n",
      "[22, 23, 24, 25, 27, 36, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 120, 121, 122, 123]\n",
      "Gene Cd209a-204: Not enough pos for 3 binding sites.\n",
      "Gene Cd209a-204: condition too harsh, loose to get better results\n",
      "[278, 279, 280, 281, 282, 283, 284, 285, 286, 288, 289, 290, 291, 292, 293, 295, 296, 297, 298, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 315, 316, 317, 318, 325, 326, 339, 340, 341, 342, 343, 345, 346, 359, 372, 373, 374, 375, 376, 377, 378, 379, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 410]\n",
      "Gene Klrb1c-204: Not enough pos for 3 binding sites.\n",
      "Gene Klrb1c-204: condition too harsh, loose to get better results\n",
      "[45, 51, 52, 53, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 107, 108, 109, 110, 162, 163, 164, 165, 167]\n",
      "Gene Gja5-202: condition too harsh, loose to get better results\n",
      "[32, 106, 180]\n"
     ]
    }
   ],
   "source": [
    "BDS_INFO[\"wanted\"] = [True] * len(BDS_INFO)\n",
    "\n",
    "# select by specifity\n",
    "gene_name_list = [_.upper() for _ in gene_list]\n",
    "gene_name_list_out = [i for i in gene_name_list]\n",
    "for i in range(len(BDS_INFO)):\n",
    "    # check gene_name\n",
    "    gene_name = BDS_INFO.loc[i, \"gene_name\"]\n",
    "    if gene_name.split('-')[0].upper() not in gene_name_list:\n",
    "        BDS_INFO.loc[i, \"wanted\"] = False\n",
    "    else:\n",
    "        try: gene_name_list_out.remove(gene_name)\n",
    "        except: pass\n",
    "\n",
    "    # check DNA or mRNA type\n",
    "    if BDS_INFO.loc[i, \"wanted\"] == True:\n",
    "        if BDS_INFO.loc[i, \"mol_type\"] != \"protein_coding\":\n",
    "            # FOI.loc[i, \"wanted\"] = False\n",
    "            print(BDS_INFO.loc[i, 'gene_name'], BDS_INFO.loc[i, \"mol_type\"])\n",
    "\n",
    "    # check gene_organism name\n",
    "    if BDS_INFO.loc[i, \"wanted\"] == True:\n",
    "        spe_ori, gene_ori = BDS_INFO.loc[i, \"organism\"], BDS_INFO.loc[i, \"gene_name\"].split('-')[0]\n",
    "        descrip = BDS_INFO.loc[i, \"align_descrip\"].split(\"|\")\n",
    "        for des in descrip:\n",
    "            if gene_ori not in des and spe_ori in des:\n",
    "                BDS_INFO.loc[i, \"wanted\"] = False\n",
    "                break\n",
    "\n",
    "    # check plus/minus\n",
    "    if BDS_INFO.loc[i, \"wanted\"] == True:\n",
    "        if pd.isnull(BDS_INFO.loc[i, \"plus/minus\"]):\n",
    "            BDS_INFO.loc[i, \"wanted\"] = False\n",
    "        else:\n",
    "            pm_list = BDS_INFO.loc[i, \"plus/minus\"].split(\",\")\n",
    "            if \"-1\" not in pm_list:\n",
    "                BDS_INFO.loc[i, \"wanted\"] = False\n",
    "\n",
    "# write the whole information of interest to a excel file in tmp dir\n",
    "BDS_INFO.to_excel(os.path.join(output, \"probes_candidates.xlsx\"))\n",
    "\n",
    "out_tmp = BDS_INFO[BDS_INFO[\"wanted\"] == True]\n",
    "output_df = pd.DataFrame()\n",
    "for gene in out_tmp.gene_name.unique():\n",
    "    pos_wanted = list(out_tmp[out_tmp.gene_name == gene][\"pos_on_seq\"])\n",
    "    pos_best = optimize_subsequence(pos_wanted, length=3, min_gap=40, better_gap=80, gene=gene)\n",
    "    pos_output = out_tmp[out_tmp.gene_name == gene]\n",
    "    pos_output = pos_output[pos_output[\"pos_on_seq\"].isin(pos_best)]\n",
    "    output_df = pd.concat([output_df, pos_output])\n",
    "\n",
    "# write the output to a xlsx file\n",
    "output_df.to_excel(os.path.join(output, \"probes_wanted.xlsx\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
