{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from pprint import pprint\n",
    "\n",
    "\n",
    "workdir = r\"../dataset/2023.11.20_mutation\"\n",
    "info = pd.read_excel(os.path.join(workdir, \"selected_mutation_list_SJT002HCC.xlsx\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get seq from ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "\n",
    "\n",
    "def padlock_thre(Tm_left, Tm_right, left, right, Tm_dif_thre=10, Tm_sing_thre=45, bind_sing_thre=10):\n",
    "    flag = False\n",
    "    if abs(Tm_left - Tm_right) < Tm_dif_thre:\n",
    "        if Tm_left > Tm_sing_thre:\n",
    "            if Tm_right > Tm_sing_thre:\n",
    "                flag = True\n",
    "            else:\n",
    "                right += 1\n",
    "        else:\n",
    "            left += 1\n",
    "    elif Tm_left > Tm_right:\n",
    "        left -= 1\n",
    "        right += 1\n",
    "        if left <= bind_sing_thre:\n",
    "            flag = True\n",
    "    else:\n",
    "        left += 1\n",
    "        right -= 1\n",
    "        if right <= bind_sing_thre:\n",
    "            flag = True\n",
    "    return left, right, flag\n",
    "\n",
    "\n",
    "def fetch_and_label_sequences(positions, gap=50, strand=1, left_length=20, right_length=20, Tm_dif_thre=10,Tm_sing_thre=45, bind_sing_thre=10):\n",
    "    # Ensembl REST API URL for GRCh37 batch sequence fetching\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    sequences_info = []\n",
    "    coord_system_version = \"GRCh37\"\n",
    "\n",
    "    for position in tqdm(positions):\n",
    "        left = left_length\n",
    "        right = right_length\n",
    "        # Adjust start and end for the extra 50 base pairs\n",
    "        chromosome = position[\"chr\"].replace(\"chr\", \"\")\n",
    "        adjusted_start = int(position[\"start\"])  # Ensure start is not less than 1\n",
    "        adjusted_end = int(position[\"end\"])\n",
    "        ext = f\"/sequence/region/human/{chromosome}:{adjusted_start}..{adjusted_end}:{strand}?\"\n",
    "\n",
    "        options = \";\".join([\n",
    "                # 'content-type=text/x-fasta',\n",
    "                \"content-type=application/json\",\n",
    "                f\"coord_system_version={coord_system_version}\",\n",
    "                f\"expand_3prime={gap}\",\n",
    "                f\"expand_5prime={gap}\",\n",
    "            ])\n",
    "\n",
    "        response = requests.get(server + ext + options)\n",
    "\n",
    "        if response.ok:\n",
    "            decoder = response.json()\n",
    "            seq = decoder[\"seq\"]\n",
    "            while True:\n",
    "                binding_left = seq[gap + 1 + adjusted_end-adjusted_start - left: gap + 1 + adjusted_end-adjusted_start]\n",
    "                binding_right = seq[-gap : -gap + right]\n",
    "                Tm_left = mt.Tm_NN(binding_left, nn_table=mt.R_DNA_NN1)\n",
    "                Tm_right = mt.Tm_NN(binding_right, nn_table=mt.R_DNA_NN1)\n",
    "                left, right, flag = padlock_thre(Tm_left, Tm_right, left, right, \n",
    "                                                 Tm_dif_thre=Tm_dif_thre, Tm_sing_thre=Tm_sing_thre, \n",
    "                                                 bind_sing_thre=bind_sing_thre)\n",
    "                if flag:\n",
    "                    break\n",
    "                \n",
    "            decoder['gene'] = position['gene']\n",
    "            decoder[\"binding_left\"] = binding_left\n",
    "            decoder[\"Tm_left\"] = Tm_left\n",
    "            decoder[\"binding_right\"] = binding_right\n",
    "            decoder[\"Tm_right\"] = Tm_right\n",
    "            sequences_info.append(decoder)\n",
    "        else:\n",
    "            response.raise_for_status()\n",
    "\n",
    "    return sequences_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/61 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 61/61 [01:28<00:00,  1.46s/it]\n"
     ]
    }
   ],
   "source": [
    "# Example usage\n",
    "positions = [\n",
    "    {\n",
    "        \"gene\": info.loc[line, \"Gene.refGene\"],\n",
    "        \"chr\": info.loc[line, \"Chr\"],\n",
    "        \"start\": info.loc[line, \"Start\"],\n",
    "        \"end\": info.loc[line, \"End\"],\n",
    "    }\n",
    "    for line in info.index\n",
    "]\n",
    "\n",
    "gap = 50\n",
    "sequences = fetch_and_label_sequences(positions, gap=gap, strand=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## perform mutation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "\n",
    "\n",
    "def binding_mutation(mutation, seq, adjusted_start, adjusted_end, left, right):\n",
    "    if mutation[\"ref\"] == \"-\":\n",
    "        mut_seq = seq[:adjusted_end] + mutation[\"alt\"] + seq[adjusted_end:]\n",
    "        add = len(mutation[\"alt\"])\n",
    "        if add < left:\n",
    "            binding_left = seq[:adjusted_end][add - left :] + mutation[\"alt\"]\n",
    "        else:\n",
    "            binding_left = mutation[\"alt\"][-left:]\n",
    "        binding_right = seq[adjusted_end:][:right]\n",
    "\n",
    "    elif mutation[\"alt\"] == \"-\":\n",
    "        mut_seq = seq[:adjusted_start] + seq[adjusted_end:]\n",
    "        binding_left = seq[:adjusted_start][-left:]\n",
    "        binding_right = seq[adjusted_end:][:right]\n",
    "\n",
    "    else:\n",
    "        mut_seq = seq[:adjusted_start] + mutation[\"alt\"] + seq[adjusted_end:]\n",
    "        add = len(mutation[\"alt\"])\n",
    "        if add < left:\n",
    "            binding_left = seq[:adjusted_start][add - left :] + mutation[\"alt\"]\n",
    "        else:\n",
    "            binding_left = mutation[\"alt\"][-left:]\n",
    "        binding_right = seq[adjusted_end:][:right]\n",
    "    return mut_seq, binding_left, binding_right\n",
    "\n",
    "\n",
    "def perform_mutation(\n",
    "    positions,\n",
    "    mutations,\n",
    "    sequences,\n",
    "    gap=50,\n",
    "    left_length=20,\n",
    "    right_length=20,\n",
    "    Tm_thre=45, \n",
    "    Tm_dif_thre=10,\n",
    "    bind_single_thre=10,\n",
    "):\n",
    "    mutation_seqs = []\n",
    "    for _ in range(len(mutations)):\n",
    "        position = positions[_]\n",
    "        mutation = mutations[_]\n",
    "        sequence = sequences[_]\n",
    "        left, right = left_length, right_length\n",
    "\n",
    "        adjusted_start = gap  # Ensure start is not less than 1\n",
    "        adjusted_end = int(position[\"end\"] - position[\"start\"] + gap + 1)\n",
    "        seq = sequence[\"seq\"]\n",
    "\n",
    "        while True:\n",
    "            mut_seq, binding_left, binding_right = binding_mutation(\n",
    "                mutation, seq, adjusted_start, adjusted_end, left, right\n",
    "            )\n",
    "            Tm_left = mt.Tm_NN(binding_left, nn_table=mt.R_DNA_NN1)\n",
    "            Tm_right = mt.Tm_NN(binding_right, nn_table=mt.R_DNA_NN1)\n",
    "            left, right, flag = padlock_thre(Tm_left, Tm_right, left, right, \n",
    "                                             Tm_dif_thre=Tm_dif_thre, Tm_sing_thre=Tm_thre, \n",
    "                                             bind_sing_thre=bind_single_thre)\n",
    "            if flag:\n",
    "                break\n",
    "\n",
    "        mutation_seq_info = {\n",
    "            \"gene\": sequence[\"gene\"],\n",
    "            \"id\": sequence[\"id\"],\n",
    "            \"molecule\": sequence[\"molecule\"],\n",
    "            \"seq\": mut_seq,\n",
    "            \"binding_left\": binding_left,\n",
    "            \"Tm_left\": Tm_left,\n",
    "            \"binding_right\": binding_right,\n",
    "            \"Tm_right\": Tm_right,\n",
    "        }\n",
    "\n",
    "        mutation_seqs.append(mutation_seq_info)\n",
    "\n",
    "    return mutation_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations = [\n",
    "    {\n",
    "        \"ref\": info.loc[line, \"Ref\"],\n",
    "        \"alt\": info.loc[line, \"Alt\"],\n",
    "    }\n",
    "    for line in info.index\n",
    "]\n",
    "\n",
    "mut_sequences = perform_mutation(\n",
    "    positions=positions, mutations=mutations, sequences=sequences, gap=50\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## generate fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "\n",
    "def generate_binding_fastq(sequences, outpath):\n",
    "    with open(outpath, \"w\") as f:\n",
    "        for sequence in sequences:\n",
    "            record = SeqRecord(\n",
    "                Seq(sequence[\"binding_left\"] + sequence[\"binding_right\"]),\n",
    "                id=sequence[\"id\"],\n",
    "                description=sequence['gene'],\n",
    "            )\n",
    "            SeqIO.write(record, f, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "generate_binding_fastq(sequences=sequences, outpath=os.path.join(workdir, 'non-mut-binding_Tm_revised.fasta'))\n",
    "generate_binding_fastq(sequences=mut_sequences, outpath=os.path.join(workdir, 'mut-binding_Tm_revised.fasta'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## blast_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "from copy import deepcopy\n",
    "\n",
    "def extract_blast(sequences, blast_results):\n",
    "    sequences_tmp = deepcopy(sequences)\n",
    "    # read the id/plus-minus part/align_num\n",
    "    with open(blast_results, \"r\") as blast_output:\n",
    "        blast_records = NCBIXML.parse(blast_output)\n",
    "        for num, blast_record in enumerate(blast_records):\n",
    "            length = len(blast_record.alignments)\n",
    "            sequences_tmp[num][\"align_num\"] = length\n",
    "            sequences_tmp[num]['descrip'] = dict()\n",
    "            for i in range(length):\n",
    "                pm = blast_record.alignments[i].hsps[0].frame[1]\n",
    "                sequences_tmp[num][\"descrip\"][i+1] = (\n",
    "                    blast_record.descriptions[i].title + f\"| pm={pm}\"\n",
    "                )\n",
    "\n",
    "    return sequences_tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_with_blast = extract_blast(sequences=sequences, blast_results=os.path.join(workdir, 'non_mut_revised.xml'))\n",
    "mut_sequences_with_blast = extract_blast(sequences=mut_sequences, blast_results=os.path.join(workdir, 'mut_revised.xml'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## perform minus for positive combined seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trans(seq):\n",
    "    translib = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n",
    "    seq_minus = \"\".join(list(reversed([translib[i] for i in seq])))\n",
    "    return seq_minus\n",
    "\n",
    "\n",
    "def specificity_judge(sequence_with_blast):\n",
    "    gene = sequence_with_blast['gene']\n",
    "    # print(gene)\n",
    "    if ';' in gene:\n",
    "        gene = list(gene.split(';'))\n",
    "    else:\n",
    "        gene = [gene]\n",
    "\n",
    "    perform_minus = False\n",
    "    specify = True\n",
    "    if len(sequence_with_blast['descrip'].values()) == 0:\n",
    "        pass\n",
    "    else:\n",
    "        for desc in sequence_with_blast['descrip'].values():\n",
    "            flag = True\n",
    "            for subgene in gene:\n",
    "                if subgene.upper() in desc.upper():\n",
    "                    flag = False\n",
    "                    pm = int(desc.split('|')[-1].split('=')[-1])\n",
    "                    if pm == 1:\n",
    "                        perform_minus = True\n",
    "            if flag:\n",
    "                specify = False\n",
    "    return perform_minus, specify\n",
    "    \n",
    "    \n",
    "def minus(sequences_with_blast, mut_sequences_with_blast):\n",
    "    sequences_final = []\n",
    "    mut_sequences_final = []\n",
    "    for _ in range(len(sequences_with_blast)):\n",
    "        seq_record = {}\n",
    "        mut_seq_record = {}\n",
    "\n",
    "        sequence_with_blast = sequences_with_blast[_]\n",
    "        mut_sequence_with_blast = mut_sequences_with_blast[_]\n",
    "\n",
    "        seq_record['gene'] = sequence_with_blast['gene']\n",
    "        mut_seq_record['gene'] = mut_sequence_with_blast['gene']\n",
    "        \n",
    "        seq_record['specificity'], perform_minus = specificity_judge(sequence_with_blast)\n",
    "        mut_seq_record['specificity'], _ = specificity_judge(mut_sequence_with_blast)\n",
    "        \n",
    "        seq_record['perform_minus'] = perform_minus\n",
    "        mut_seq_record['perform_minus'] = perform_minus\n",
    " \n",
    "        if perform_minus:\n",
    "            seq_record['binding_left'] = trans(sequence_with_blast['binding_right'])\n",
    "            seq_record['binding_right'] = trans(sequence_with_blast['binding_left'])\n",
    "            seq_record['Tm_left'] = sequence_with_blast['Tm_right']\n",
    "            seq_record['Tm_right'] = sequence_with_blast['Tm_left']\n",
    "            seq_record['seq'] = sequence_with_blast['seq']\n",
    "            \n",
    "            mut_seq_record['binding_left'] = trans(mut_sequence_with_blast['binding_right'])\n",
    "            mut_seq_record['binding_right'] = trans(mut_sequence_with_blast['binding_left'])\n",
    "            mut_seq_record['Tm_left'] = mut_sequence_with_blast['Tm_right']\n",
    "            mut_seq_record['Tm_right'] = mut_sequence_with_blast['Tm_left']\n",
    "            mut_seq_record['seq'] = mut_sequence_with_blast['seq']\n",
    "\n",
    "        else:\n",
    "            seq_record['binding_left'] = sequence_with_blast['binding_left']\n",
    "            seq_record['binding_right'] = sequence_with_blast['binding_right']\n",
    "            seq_record['Tm_left'] = sequence_with_blast['Tm_left']\n",
    "            seq_record['Tm_right'] = sequence_with_blast['Tm_right']\n",
    "            seq_record['seq'] = sequence_with_blast['seq']\n",
    "            \n",
    "            mut_seq_record['binding_left'] = mut_sequence_with_blast['binding_left']\n",
    "            mut_seq_record['binding_right'] = mut_sequence_with_blast['binding_right']\n",
    "            mut_seq_record['Tm_left'] = mut_sequence_with_blast['Tm_left']\n",
    "            mut_seq_record['Tm_right'] = mut_sequence_with_blast['Tm_right']\n",
    "            mut_seq_record['seq'] = mut_sequence_with_blast['seq']\n",
    "            \n",
    "        sequences_final.append(seq_record)\n",
    "        mut_sequences_final.append(mut_seq_record)\n",
    "    \n",
    "    return sequences_final, mut_sequences_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_final, mut_sequences_final = minus(sequences_with_blast, mut_sequences_with_blast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "interest = ['gene', 'binding_left','binding_right','Tm_left','Tm_right','specificity','perform_minus']\n",
    "\n",
    "df_dict = dict()\n",
    "for inte in interest:\n",
    "    df_dict[inte] = []\n",
    "\n",
    "for _ in range(len(sequences_final)):\n",
    "    sequence = sequences_final[_]\n",
    "    mut_sequence = mut_sequences_final[_]\n",
    "    for inte in interest:\n",
    "        if inte == 'gene':\n",
    "            df_dict[inte] += [sequence[inte], mut_sequence[inte]+'_mut']\n",
    "        else:\n",
    "            df_dict[inte] += [sequence[inte], mut_sequence[inte]]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>binding</th>\n",
       "      <th>binding_left</th>\n",
       "      <th>binding_right</th>\n",
       "      <th>Tm_left</th>\n",
       "      <th>Tm_right</th>\n",
       "      <th>specificity</th>\n",
       "      <th>perform_minus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COL1A2</td>\n",
       "      <td>TCTACAAGAACAGTGTAAGTGAACCTGCTGTTGCCCTCAG</td>\n",
       "      <td>TCTACAAGAACAGTGTAAGTG</td>\n",
       "      <td>AACCTGCTGTTGCCCTCAG</td>\n",
       "      <td>46.356676</td>\n",
       "      <td>52.424935</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COL1A2_mut</td>\n",
       "      <td>TCTACAAGAACAGTGTAAGTGTACCTGCTGTTGCCCTCAG</td>\n",
       "      <td>TCTACAAGAACAGTGTAAGTG</td>\n",
       "      <td>TACCTGCTGTTGCCCTCAG</td>\n",
       "      <td>46.356676</td>\n",
       "      <td>51.532012</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>INTS1</td>\n",
       "      <td>CCCAGCGGGAGAAGCAGGAGATCCTGGCCTTCGAGGGGCA</td>\n",
       "      <td>CCCAGCGGGAGAAGCAGGAG</td>\n",
       "      <td>ATCCTGGCCTTCGAGGGGCA</td>\n",
       "      <td>66.053131</td>\n",
       "      <td>61.566136</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>INTS1_mut</td>\n",
       "      <td>CCCAGCGGGAGAAGCAGGAGTTCCTGGCCTTCGAGGGGCA</td>\n",
       "      <td>CCCAGCGGGAGAAGCAGGAG</td>\n",
       "      <td>TTCCTGGCCTTCGAGGGGCA</td>\n",
       "      <td>66.053131</td>\n",
       "      <td>59.827105</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HTR3E</td>\n",
       "      <td>CCCCCTGTCAGCTCTGCCTCCGCAGGGCCCGGCATCTGCC</td>\n",
       "      <td>CCCCCTGTCAGCTCTGCCTCC</td>\n",
       "      <td>GCAGGGCCCGGCATCTGCC</td>\n",
       "      <td>60.309166</td>\n",
       "      <td>67.353106</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>ESR1_mut</td>\n",
       "      <td>GGCTAGTGGGCGCATGTAGGTGGTGGGCGTCCAGCATCTC</td>\n",
       "      <td>GGCTAGTGGGCGCATGTAGG</td>\n",
       "      <td>TGGTGGGCGTCCAGCATCTC</td>\n",
       "      <td>61.513585</td>\n",
       "      <td>59.512869</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>FTL</td>\n",
       "      <td>CAATTCGCGGAAGAAGTGGCTCACGCCTTCCAGAGCCACA</td>\n",
       "      <td>CAATTCGCGGAAGAAGTGGC</td>\n",
       "      <td>TCACGCCTTCCAGAGCCACA</td>\n",
       "      <td>53.721559</td>\n",
       "      <td>55.676675</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>FTL_mut</td>\n",
       "      <td>CAATTCGCGGAAGAAGTGGCCTACGCCTTCCAGAGCCACA</td>\n",
       "      <td>CAATTCGCGGAAGAAGTGGC</td>\n",
       "      <td>CTACGCCTTCCAGAGCCACA</td>\n",
       "      <td>53.721559</td>\n",
       "      <td>54.079431</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>KCNJ15</td>\n",
       "      <td>GCTTCACCAGGGGGGTGCTGGACATGCCGATGTGAATGGC</td>\n",
       "      <td>GCTTCACCAGGGGGGTGCTG</td>\n",
       "      <td>GACATGCCGATGTGAATGGC</td>\n",
       "      <td>64.047839</td>\n",
       "      <td>55.494756</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>KCNJ15_mut</td>\n",
       "      <td>GCTTCACCAGGGGGGTGCTGCACATGCCGATGTGAATGGC</td>\n",
       "      <td>GCTTCACCAGGGGGGTGCTG</td>\n",
       "      <td>CACATGCCGATGTGAATGGC</td>\n",
       "      <td>64.047839</td>\n",
       "      <td>54.256746</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene                                   binding  \\\n",
       "0        COL1A2  TCTACAAGAACAGTGTAAGTGAACCTGCTGTTGCCCTCAG   \n",
       "1    COL1A2_mut  TCTACAAGAACAGTGTAAGTGTACCTGCTGTTGCCCTCAG   \n",
       "2         INTS1  CCCAGCGGGAGAAGCAGGAGATCCTGGCCTTCGAGGGGCA   \n",
       "3     INTS1_mut  CCCAGCGGGAGAAGCAGGAGTTCCTGGCCTTCGAGGGGCA   \n",
       "4         HTR3E  CCCCCTGTCAGCTCTGCCTCCGCAGGGCCCGGCATCTGCC   \n",
       "..          ...                                       ...   \n",
       "117    ESR1_mut  GGCTAGTGGGCGCATGTAGGTGGTGGGCGTCCAGCATCTC   \n",
       "118         FTL  CAATTCGCGGAAGAAGTGGCTCACGCCTTCCAGAGCCACA   \n",
       "119     FTL_mut  CAATTCGCGGAAGAAGTGGCCTACGCCTTCCAGAGCCACA   \n",
       "120      KCNJ15  GCTTCACCAGGGGGGTGCTGGACATGCCGATGTGAATGGC   \n",
       "121  KCNJ15_mut  GCTTCACCAGGGGGGTGCTGCACATGCCGATGTGAATGGC   \n",
       "\n",
       "              binding_left         binding_right    Tm_left   Tm_right  \\\n",
       "0    TCTACAAGAACAGTGTAAGTG   AACCTGCTGTTGCCCTCAG  46.356676  52.424935   \n",
       "1    TCTACAAGAACAGTGTAAGTG   TACCTGCTGTTGCCCTCAG  46.356676  51.532012   \n",
       "2     CCCAGCGGGAGAAGCAGGAG  ATCCTGGCCTTCGAGGGGCA  66.053131  61.566136   \n",
       "3     CCCAGCGGGAGAAGCAGGAG  TTCCTGGCCTTCGAGGGGCA  66.053131  59.827105   \n",
       "4    CCCCCTGTCAGCTCTGCCTCC   GCAGGGCCCGGCATCTGCC  60.309166  67.353106   \n",
       "..                     ...                   ...        ...        ...   \n",
       "117   GGCTAGTGGGCGCATGTAGG  TGGTGGGCGTCCAGCATCTC  61.513585  59.512869   \n",
       "118   CAATTCGCGGAAGAAGTGGC  TCACGCCTTCCAGAGCCACA  53.721559  55.676675   \n",
       "119   CAATTCGCGGAAGAAGTGGC  CTACGCCTTCCAGAGCCACA  53.721559  54.079431   \n",
       "120   GCTTCACCAGGGGGGTGCTG  GACATGCCGATGTGAATGGC  64.047839  55.494756   \n",
       "121   GCTTCACCAGGGGGGTGCTG  CACATGCCGATGTGAATGGC  64.047839  54.256746   \n",
       "\n",
       "     specificity  perform_minus  \n",
       "0           True           True  \n",
       "1          False           True  \n",
       "2          False           True  \n",
       "3          False           True  \n",
       "4           True           True  \n",
       "..           ...            ...  \n",
       "117        False           True  \n",
       "118         True           True  \n",
       "119        False           True  \n",
       "120         True           True  \n",
       "121        False           True  \n",
       "\n",
       "[122 rows x 8 columns]"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(df_dict)\n",
    "df['binding'] = df['binding_left'] + df['binding_right']\n",
    "df = df[['gene', 'binding', 'binding_left','binding_right','Tm_left','Tm_right','specificity','perform_minus']]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'gene': 'COL1A2',\n",
       "  'id': 'chromosome:GRCh37:7:94058661:94058761:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TCTACAGGGCTCTAATGATGTTGAACTTGTTGCTGAGGGCAACAGCAGGTACACTTACACTGTTCTTGTAGATGGCTGCTCTGTAAGTAATAGTGAAATAT',\n",
       "  'binding_left': 'CTGAGGGCAACAGCAGGTA',\n",
       "  'Tm_left': 51.53201200896723,\n",
       "  'binding_right': 'CACTTACACTGTTCTTGTAGA',\n",
       "  'Tm_right': 46.35667628475903,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'INTS1',\n",
       "  'id': 'chromosome:GRCh37:7:1527446:1527546:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GATGGTCTGCTTGGTGGACGCGGCCGCCAGGTGCCCCTCGAAGGCCAGGAACTCCTGCTTCTCCCGCTGGGCGGTCTGCAGCTCACGGTTCAGCATCTCCG',\n",
       "  'binding_left': 'TGCCCCTCGAAGGCCAGGAA',\n",
       "  'Tm_left': 59.82710529545699,\n",
       "  'binding_right': 'CTCCTGCTTCTCCCGCTGGG',\n",
       "  'Tm_right': 66.05313092483942,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'HTR3E',\n",
       "  'id': 'chromosome:GRCh37:3:183824247:183824347:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TGTAGGTGTGAAGGAGCCAGAGGTATCAGCAGGGCAGATGCCGGGCCCTGGGGAGGCAGAGCTGACAGGGGGCTCAGAATGGACAAGGGCCCAGCGGGAAC',\n",
       "  'binding_left': 'GGGCAGATGCCGGGCCCTGG',\n",
       "  'Tm_left': 67.81069581147506,\n",
       "  'binding_right': 'GGAGGCAGAGCTGACAGGGG',\n",
       "  'Tm_right': 57.85155207521393,\n",
       "  'align_num': 5,\n",
       "  'descrip': {1: 'gi|1676317206|ref|NM_182537.3| Homo sapiens 5-hydroxytryptamine receptor 3D (HTR3D), transcript variant 2, mRNA| pm=1',\n",
       "   2: 'gi|1675144830|ref|NM_001163646.2| Homo sapiens 5-hydroxytryptamine receptor 3D (HTR3D), transcript variant 3, mRNA| pm=1',\n",
       "   3: 'gi|2462587945|ref|XM_054345628.1| PREDICTED: Homo sapiens 5-hydroxytryptamine receptor 3D (HTR3D), transcript variant X1, mRNA| pm=1',\n",
       "   4: 'gi|2287254618|ref|NM_001410851.1| Homo sapiens 5-hydroxytryptamine receptor 3D (HTR3D), transcript variant 4, mRNA| pm=1',\n",
       "   5: 'gi|223468704|ref|NM_001145143.1| Homo sapiens 5-hydroxytryptamine receptor 3D (HTR3D), transcript variant 1, mRNA| pm=1'}},\n",
       " {'gene': 'NDUFA11',\n",
       "  'id': 'chromosome:GRCh37:19:5893050:5893150:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'AGGCCCGGGCACCCAGCTCCGGAAGTGGACGTGACCCTGAGCTGTTGCATGCCCCCCCTTGTGCTGGAACACGCGTGGACATCTGTGCTCTCTCCAGAGAGG',\n",
       "  'binding_left': 'GACCCTGAGCTGTTGCATGC',\n",
       "  'Tm_left': 58.71679576408468,\n",
       "  'binding_right': 'CCCCCCTTGTGCTGGAACAC',\n",
       "  'Tm_right': 61.38998345983947,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'COQ8A',\n",
       "  'id': 'chromosome:GRCh37:1:227149163:227149263:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'ACCTGCAGCACTTGGGCATCGGAGGGGAGCTGATCATGGCGGCCAGGGCCGTGCAGTCCACGGCTGTGGAGCAGATTGGCATGTTCTTGGGGAAGGTGCAG',\n",
       "  'binding_left': 'GATCATGGCGGCCAGGGCCG',\n",
       "  'Tm_left': 63.27194962892639,\n",
       "  'binding_right': 'TGCAGTCCACGGCTGTGGAG',\n",
       "  'Tm_right': 58.850101227690345,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'OR52I1',\n",
       "  'id': 'chromosome:GRCh37:11:4616107:4616207:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TGCTAGCTGACCTGTACGTGATCATCCCAGCCACTTTAAATCCCATCATCCATGGCATGAGGACCAAACAATTGCTGGAGGGAATATGGAGTTATCTGATG',\n",
       "  'binding_left': 'CACTTTAAATCCCATCATCC',\n",
       "  'Tm_left': 46.69880786423778,\n",
       "  'binding_right': 'ATGGCATGAGGACCAAACAATT',\n",
       "  'Tm_right': 45.69273727017128,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'GAB4',\n",
       "  'id': 'chromosome:GRCh37:22:17450917:17451017:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTCTGAATTCTGCATTGTGCTGGCTGGGCTTGGAAAGGCTATGGAAGCCACGGATGTGACCGCTGACCCCATCAACACTGTATCCACTGTGCTGGGCAAGA',\n",
       "  'binding_left': 'TTGGAAAGGCTATGGAAGCCAC',\n",
       "  'Tm_left': 49.46382137840129,\n",
       "  'binding_right': 'GGATGTGACCGCTGACCC',\n",
       "  'Tm_right': 55.845655408656626,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'HPX',\n",
       "  'id': 'chromosome:GRCh37:11:6453081:6453181:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CCAGTCTCTACCTCAAGCCTCTCCCCCAATACACACCTGGACCAGATAGAGTTTTTTCTTCCCAGGAAAAGGCAGCATCCACTGCTGAAGGACCCTGGGGCC',\n",
       "  'binding_left': 'ACACCTGGACCAGATAGAGT',\n",
       "  'Tm_left': 50.899146599759035,\n",
       "  'binding_right': 'TTTTTCTTCCCAGGAAAAGGC',\n",
       "  'Tm_right': 48.89208048477701,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'DCC',\n",
       "  'id': 'chromosome:GRCh37:18:51013152:51013252:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTTTTTCTTTTGCAGCTGTCGTGAGCGCCATCCCGGTGCCAACGCTAGAAGGTGCCCAGTACCCAGGAATCCTCCCGTCTCCCACCTGTGGATATCCCCAC',\n",
       "  'binding_left': 'CCCGGTGCCAACGCTAGAAG',\n",
       "  'Tm_left': 55.282960036706356,\n",
       "  'binding_right': 'GTGCCCAGTACCCAGGAATC',\n",
       "  'Tm_right': 58.76272210142963,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'SSC4D',\n",
       "  'id': 'chromosome:GRCh37:7:76024546:76024646:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CGCCTGGAGGGCTGAGCTCTTTACCGGACGGATCTGTCTGCCAAGCCCAGCCCTCTCTTGTGGCTGAGGATGGCAGTGCTGTGAGCGTTGGGGGACCCAGG',\n",
       "  'binding_left': 'ATCTGTCTGCCAAGCCCAGC',\n",
       "  'Tm_left': 61.2405359452369,\n",
       "  'binding_right': 'CCTCTCTTGTGGCTGAGGAT',\n",
       "  'Tm_right': 56.70282861564135,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'APOBEC2',\n",
       "  'id': 'chromosome:GRCh37:6:41029289:41029389:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TGTCACCTGGTATGTGTCCTCCAGCCCCTGTGCAGCGTGTGCTGACCGCACTATCAAAACCCTTAGCAAGACCAAGAACCTGCGTCTGCTCATTCTGGTGG',\n",
       "  'binding_left': 'AGCGTGTGCTGACCGCAC',\n",
       "  'Tm_left': 56.12865642712575,\n",
       "  'binding_right': 'TATCAAAACCCTTAGCAAGACC',\n",
       "  'Tm_right': 46.20566466446638,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'DRP2',\n",
       "  'id': 'chromosome:GRCh37:X:100490851:100490951:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TAGAGCTGCTGTCACCAGCCCTGCACCTCCTCAAGATGGTGCTGGGGTTCACTGCCTAAGCCTAAAGCTGTTGAACGGGTCTGTTGGTGCCTCTGGACCCC',\n",
       "  'binding_left': 'CAAGATGGTGCTGGGGTTCA',\n",
       "  'Tm_left': 53.20758949200871,\n",
       "  'binding_right': 'CTGCCTAAGCCTAAAGCTGT',\n",
       "  'Tm_right': 51.80937233872231,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'TMTC4',\n",
       "  'id': 'chromosome:GRCh37:13:101287293:101287393:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TTTGCTCAGGGCTCCGAATCCAAAAGTCAGCAGCACACAGTACCCAACGCAGGGGAGGTAGAGGACACGCTCTGCGACCACGAAGCCCACTCGGAAGAACA',\n",
       "  'binding_left': 'AGCACACAGTACCCAACGCA',\n",
       "  'Tm_left': 53.70686100409546,\n",
       "  'binding_right': 'GGGGAGGTAGAGGACACGCT',\n",
       "  'Tm_right': 56.11033530889159,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'KCNA5',\n",
       "  'id': 'chromosome:GRCh37:12:5154816:5154916:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CATCGCCGGGGTCCTCACCATTGCCCTGCCTGTGCCCGTCATCGTCTCCATCTTCAACTACTTCTACCACCGGGAAACGGATCACGAGGAGCCGGCAGTCC',\n",
       "  'binding_left': 'GCCCGTCATCGTCTCCAT',\n",
       "  'Tm_left': 59.97182883438006,\n",
       "  'binding_right': 'CTTCAACTACTTCTACCACCGG',\n",
       "  'Tm_right': 52.299583333956036,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'MAGEB3',\n",
       "  'id': 'chromosome:GRCh37:X:30254124:30254224:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTCAGATCACTGCAACTAACAAGAAAAAAGTATCCTTTTCATCCCCTCTTGTTTTGGGGGCTACTATCCAGAAAAAGTCTGCTGGTAGGTCACGTAGTGCT',\n",
       "  'binding_left': 'ATCCTTTTCATCCCCTCTTG',\n",
       "  'Tm_left': 56.46920844066739,\n",
       "  'binding_right': 'TTTTGGGGGCTACTATCCAG',\n",
       "  'Tm_right': 52.63969847688321,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'DNAJC1',\n",
       "  'id': 'chromosome:GRCh37:10:22048062:22048162:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TAACACCTGTGCAGAGGCCTCTTTCCCCATCACTCACCTTGCTCTTGGACAGGACACATCTGGCTATTTTGTCCCAGCGGTCAGAGGATCCCCTTGGGTAC',\n",
       "  'binding_left': 'ACTCACCTTGCTCTTGGACA',\n",
       "  'Tm_left': 54.318105687663945,\n",
       "  'binding_right': 'GGACACATCTGGCTATTTTG',\n",
       "  'Tm_right': 47.29787569102069,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'ARSF',\n",
       "  'id': 'chromosome:GRCh37:X:3030246:3030346:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CCCGGTATTCCAGCCACCAGCTTCTGGTGGCTGCTATGTCACCTCATTATACAGATGTTTCGGAGAACAGGTTACCTACCACAACCCCCCTCTGCTCTTCG',\n",
       "  'binding_left': 'TGCTATGTCACCTCATTATA',\n",
       "  'Tm_left': 46.57762294292718,\n",
       "  'binding_right': 'CAGATGTTTCGGAGAACAGGT',\n",
       "  'Tm_right': 47.72855700165093,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'NDUFAF3',\n",
       "  'id': 'chromosome:GRCh37:3:49060491:49060591:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CAATGCCTGTGCCACCTTCAACTTCCTGTGTCATGAAGGCCGAGTAACTGCAGCTGCTCTCATCCCTCCACCAGGAGGGACTTCACTTACATCTTTGGGCC',\n",
       "  'binding_left': 'GTCATGAAGGCCGAGTAACTGC',\n",
       "  'Tm_left': 52.96590784816914,\n",
       "  'binding_right': 'AGCTGCTCTCATCCCTCC',\n",
       "  'Tm_right': 61.7167535881639,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'PDK2',\n",
       "  'id': 'chromosome:GRCh37:17:48185696:48185796:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CGACTGTGGAAAGCCATGAGTCCAGCCTCATTCTCCCACCCATCAAGGTCGTGGTGGCCTTGGGTGAGGAAGATCTGTCCATCAAGGTATGTGACCCTTTG',\n",
       "  'binding_left': 'TCTCCCACCCATCAAGGTCG',\n",
       "  'Tm_left': 58.68651073521329,\n",
       "  'binding_right': 'TGGTGGCCTTGGGTGAGGAA',\n",
       "  'Tm_right': 56.92855803726832,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'DOCK4',\n",
       "  'id': 'chromosome:GRCh37:7:111379223:111379323:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'ACTGCATGGTCTCTCTCTTGGTGACAGGCAAGAGTTTTCTCGGGAATGTTCGTGTTTGTCAGACAACAAAGGAGAAGCTGGGAAGAGAAGGAGAGAGAGGA',\n",
       "  'binding_left': 'GAGTTTTCTCGGGAATGTTC',\n",
       "  'Tm_left': 47.19501683233659,\n",
       "  'binding_right': 'GTGTTTGTCAGACAACAAAGGAG',\n",
       "  'Tm_right': 45.54568013304953,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'UTP14A',\n",
       "  'id': 'chromosome:GRCh37:X:129063466:129063566:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'AAGCAGAAGACGTGGGCTACCGGTCTTCCTCAAGGTCGGACCTGTCTGTCGTACAGAGGAATCCAAAACGAATCACCACACGTCACAAAAAACAGCTGAAG',\n",
       "  'binding_left': 'TCGGACCTGTCTGTCG',\n",
       "  'Tm_left': 49.255663408515375,\n",
       "  'binding_right': 'TACAGAGGAATCCAAAACGAATCACC',\n",
       "  'Tm_right': 46.65784784500579,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'OR5L1',\n",
       "  'id': 'chromosome:GRCh37:11:55579773:55579887:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTACACAGTCGTGATTCCTATGCTGAACTCTGTGATCTACAGCCTGAGAAAAGCTCTCAGAAAAGTGATGGGCTCCAAAATTCACTCCTAGGGAAGATTT',\n",
       "  'binding_left': 'TGTGATCTACAGCCTGAGAA',\n",
       "  'Tm_left': 47.31649006651253,\n",
       "  'binding_right': 'AAGCTCTCAGAAAAGTGATGGGC',\n",
       "  'Tm_right': 48.747222372243925,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'NUP155',\n",
       "  'id': 'chromosome:GRCh37:5:37303418:37303518:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTTCTCGTAATACCGCCAGAGTAAATCCATATAACGAACTCTGTTTTGATAAACTTTGGCCATTCGGACTAGATGTGGCTCCAGAAATGGAGAAGCAACCT',\n",
       "  'binding_left': 'CCATATAACGAACTCTGTTTTGATA',\n",
       "  'Tm_left': 45.39821623586391,\n",
       "  'binding_right': 'AACTTTGGCCATTCGGAC',\n",
       "  'Tm_right': 46.91612004761305,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'ARHGAP23',\n",
       "  'id': 'chromosome:GRCh37:17:36619572:36619672:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TGACACTCTGGAGCTGTCTATCATGCCCAAGGACGAGGACATCCTCCAGCCGGTGAGTCCAGCCCCTGTGGCCTGAGAGGAGACCCCTGAGCCCTGGGGCA',\n",
       "  'binding_left': 'GACGAGGACATCCTCCAGCC',\n",
       "  'Tm_left': 58.59791475908878,\n",
       "  'binding_right': 'GGTGAGTCCAGCCCCTGTGG',\n",
       "  'Tm_right': 66.64477648846491,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'TMEM200C',\n",
       "  'id': 'chromosome:GRCh37:18:5890209:5890309:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GCTTTCCAGTTCTTCTTCTACCCCTATGGCATGAGACCTGGAAATCATGAGGAGTTTCTCCTTGTTTGTAAACTGCCTCTGCACCGGCTGAGGTTGCTCGG',\n",
       "  'binding_left': 'GCATGAGACCTGGAAATCATGAG',\n",
       "  'Tm_left': 48.28991616584784,\n",
       "  'binding_right': 'GAGTTTCTCCTTGTTTGTAA',\n",
       "  'Tm_right': 46.923778894603856,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'TRIM24',\n",
       "  'id': 'chromosome:GRCh37:7:138235831:138235940:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CCTTCGTGCAAGGTGTGATGCATCCCCAGTGACCAACAACACCATCCAATTCCTAGTTTCTGGGCTCAAAATATCATCAACTTAGGTGGGCCATTACCAT',\n",
       "  'binding_left': 'GACCAACAACACCATCCAAT',\n",
       "  'Tm_left': 46.46712795151069,\n",
       "  'binding_right': 'TCCTAGTTTCTGGGCTCAAA',\n",
       "  'Tm_right': 50.93038827694937,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'NR2F2',\n",
       "  'id': 'chromosome:GRCh37:15:96877352:96877453:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TTCGCGCTGACCAACGGGGATCCCCTCAACTGCCACTCGTACCTGTCCGGATATTTCCCTGCTGTTGCGCGCGGAGCCCTATCCCACGTCGCGCTTCGGC',\n",
       "  'binding_left': 'TGCCACTCGTACCTGTCCGG',\n",
       "  'Tm_left': 60.677202327174086,\n",
       "  'binding_right': 'ATATTTCCCTGCTGTTGCGC',\n",
       "  'Tm_right': 56.39689272625037,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'DCAF8L1',\n",
       "  'id': 'chromosome:GRCh37:X:27999319:27999419:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTGTGGCTGCCATTTCAATGTCTGAGGAGGCCGCCGTCACCGCTGCTACTACAGACTGCTCCTCTGGGCTGCTGAACAGGCTTTCAGTCACTAAGTCTGGT',\n",
       "  'binding_left': 'CGCCGTCACCGCTGCTACTA',\n",
       "  'Tm_left': 60.12078667213615,\n",
       "  'binding_right': 'CAGACTGCTCCTCTGGGCTG',\n",
       "  'Tm_right': 59.839253670442986,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'ANKRD30BL',\n",
       "  'id': 'chromosome:GRCh37:2:132919072:132919172:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'ATCTCTTATGTTCAGGTCCATTGTCGTCTTCTTCATCATCCTCTCCAGCTCCCAGGCTTGGCCCCGGGAGGCAGCTTTGTGGATCTTCCTGAGATCCCCAT',\n",
       "  'binding_left': 'CTTCATCATCCTCTCCAGCTC',\n",
       "  'Tm_left': 59.16352400485397,\n",
       "  'binding_right': 'CCAGGCTTGGCCCCGGGAG',\n",
       "  'Tm_right': 66.10017392809755,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'NPW',\n",
       "  'id': 'chromosome:GRCh37:16:2070049:2070149:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TGGCCTGCTCATGGGGCTGCGTCGCTCACCCTATCTGTGGCGCCGCGCGCAGCGCGCGGCCGCCGGGCCCCTGGCCAGGGACACCCTCTCCCCCGAACCCG',\n",
       "  'binding_left': 'CCTATCTGTGGCGCCGCGCGCA',\n",
       "  'Tm_left': 66.74235092121478,\n",
       "  'binding_right': 'GCGCGCGGCCGCCGGGCC',\n",
       "  'Tm_right': 71.74517323115094,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'TP53',\n",
       "  'id': 'chromosome:GRCh37:17:7577071:7577171:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGGACAGGCACAAACACACACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATAGGAA',\n",
       "  'binding_left': 'CCAGGACAGGCACAAACACA',\n",
       "  'Tm_left': 49.11608819749404,\n",
       "  'binding_right': 'CACCTCAAAGCTGTTCCGTC',\n",
       "  'Tm_right': 53.57783718869143,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'DLX6',\n",
       "  'id': 'chromosome:GRCh37:7:96636996:96637096:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TGGAAAAGGGAAAAAGATTCGGAAGCCTCGGACCATTTATTCCAGCCTGCTGCTCCAGGCTTTAAACCATCGCTTTCAGCAGACACAGTATCTGGCCCTTC',\n",
       "  'binding_left': 'ACCATTTATTCCAGCCTGCT',\n",
       "  'Tm_left': 55.95104287264064,\n",
       "  'binding_right': 'GCTCCAGGCTTTAAACCATC',\n",
       "  'Tm_right': 52.136644910006794,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'ATP23',\n",
       "  'id': 'chromosome:GRCh37:12:58335486:58335604:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TGGCGGGAGCTCCGGACGAGCGCCGGCGGGGCCCCGCGGCAGGGGAGCAGCTTGCCAGGTCTTCCCCGAGCGTCTGGCCCAGGGGAATCCCCAGCAAGGG',\n",
       "  'binding_left': 'GCCCCGCGGCAGGGGAGCAG',\n",
       "  'Tm_left': 69.37750337562841,\n",
       "  'binding_right': 'CTTGCCAGGTCTTCCCCGAG',\n",
       "  'Tm_right': 62.666471459906916,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'CHD5',\n",
       "  'id': 'chromosome:GRCh37:1:6169969:6170069:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CGGGGTGGTTGGGGTCCTGCGTCATGTTCAGGTACGCGGCCCTCCGGAGCAGCTCCTCAATGACCAACGCCTGCTCCAGCAGCTGGGGGCGGGGCGCAGCG',\n",
       "  'binding_left': 'TACGCGGCCCTCCGGAGCA',\n",
       "  'Tm_left': 63.34844219786487,\n",
       "  'binding_right': 'GCTCCTCAATGACCAACGCCT',\n",
       "  'Tm_right': 57.25984961899536,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'C1orf112',\n",
       "  'id': 'chromosome:GRCh37:1:169806124:169806224:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'ACATATTTCCTTCCAGGCGTTACCTCCTGAGCTTAGGGAACAAACTGTCCGTGAGGTCACCACAGTAGGCACTGCAGAATGCAGGAAATGGCTGAGCAGGA',\n",
       "  'binding_left': 'GCTTAGGGAACAAACTGTCCG',\n",
       "  'Tm_left': 48.69915415958167,\n",
       "  'binding_right': 'TGAGGTCACCACAGTAGGC',\n",
       "  'Tm_right': 53.48709847441535,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'DOCK2',\n",
       "  'id': 'chromosome:GRCh37:5:169503038:169503138:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TTCAAGAACCTGAAAATGAAGGTGGAGAAGGAGTACGGTGTCCGAGAGATAGTATGGGTGGTTCCTATGGCTTGGAGGGGTGGGATTGTGCAGGGTGGGGC',\n",
       "  'binding_left': 'AGTACGGTGTCCGAGAGATA',\n",
       "  'Tm_left': 49.65057979171394,\n",
       "  'binding_right': 'GTATGGGTGGTTCCTATGGC',\n",
       "  'Tm_right': 56.71300298874672,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'HTRA1',\n",
       "  'id': 'chromosome:GRCh37:10:124221530:124221630:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GCGACGCCAACACCTACGCCAACCTGTGCCAGCTGCGCGCCGCCAGCCGCAGCTCCGAGAGGCTGCACCGGCCGCCGGTCATCGTCCTGCAGCGCGGAGCC',\n",
       "  'binding_left': 'GCTGCGCGCCGCCAGCCGCA',\n",
       "  'Tm_left': 71.40642603105493,\n",
       "  'binding_right': 'GCTCCGAGAGGCTGCACCGG',\n",
       "  'Tm_right': 62.96593598690157,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'MIDEAS',\n",
       "  'id': 'chromosome:GRCh37:14:74186081:74186181:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TCTCTGAAAAAGGTAGTGCCCTTCGTGACTTCCCTGTCCCTTCCCTTGGCATCTCCGAGGCCTGGCCACCGGCAGACCCAGGGGCGTTGGACTCGTGGCTC',\n",
       "  'binding_left': 'CCCTGTCCCTTCCCTTGGCA',\n",
       "  'Tm_left': 69.22692600549749,\n",
       "  'binding_right': 'TCTCCGAGGCCTGGCCACCG',\n",
       "  'Tm_right': 65.81496715349925,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'IL1RAPL1',\n",
       "  'id': 'chromosome:GRCh37:X:29301158:29301258:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'AAATTACTCCCTTGCCCAAAGTGCTGGACTCAGTTTGATGTGGTACAAAACTTCTGGTCCTGGAGACTTTGAAGAGCCAATAGCCTTTGACGGAAGTAGAA',\n",
       "  'binding_left': 'ACTCAGTTTGATGTGGTACAAAAC',\n",
       "  'Tm_left': 45.302188510069584,\n",
       "  'binding_right': 'TTCTGGTCCTGGAGACTT',\n",
       "  'Tm_right': 50.803633393985876,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'RASGRF2',\n",
       "  'id': 'chromosome:GRCh37:5:80381679:80381779:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TGGAAAGGAAAAGCCTGGAGTTTGCCAAATCAAAGCTAGAGGAACTATCCGGGTATGCCAAGAACTTATAACAAAGGCTTGGGAGAGGAAAGCAGCAAGTT',\n",
       "  'binding_left': 'TCAAAGCTAGAGGAACTATCCG',\n",
       "  'Tm_left': 45.84976543555774,\n",
       "  'binding_right': 'GGTATGCCAAGAACTTATAACAAAGG',\n",
       "  'Tm_right': 45.54282610085215,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'CCDC22',\n",
       "  'id': 'chromosome:GRCh37:X:49105641:49105741:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CGAAGGAGCTTCAGAAGGAAATCAACTCCCTATCTGGGAAGCTGGACCGGGCGTTTGCGGTGACTGATGAGCTTGTGTTCAAGGTGTGGGGCAGGTTGGGC',\n",
       "  'binding_left': 'ATCTGGGAAGCTGGACCGGG',\n",
       "  'Tm_left': 55.42926987713025,\n",
       "  'binding_right': 'CGTTTGCGGTGACTGATGAG',\n",
       "  'Tm_right': 49.08415694824379,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'MIPOL1',\n",
       "  'id': 'chromosome:GRCh37:14:37736260:37736360:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TTAATGAAATAACATGTGAGAACACAGAATGGCCAGGGCAGAGATCAACGGATTTTCAGATCATCAGTTCTTATCCAGATGATGAGTCTGTTTACTGCACT',\n",
       "  'binding_left': 'CCAGGGCAGAGATCAACGG',\n",
       "  'Tm_left': 50.85562552399671,\n",
       "  'binding_right': 'ATTTTCAGATCATCAGTTCTTAT',\n",
       "  'Tm_right': 45.582193984092726,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'ATP11C',\n",
       "  'id': 'chromosome:GRCh37:X:138884394:138884494:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTTGTAGAGGTCAGGTTGAGGCTGTTCACATTCAATTGCTGCTCGGAGGGCATCGATGGATTCTGCTGTACACAGTGCAATGGTATCACGTACTGCATAAT',\n",
       "  'binding_left': 'TCAATTGCTGCTCGGAGGGC',\n",
       "  'Tm_left': 55.52518932116783,\n",
       "  'binding_right': 'ATCGATGGATTCTGCTGTAC',\n",
       "  'Tm_right': 47.24120053855836,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'DOCK11',\n",
       "  'id': 'chromosome:GRCh37:X:117744211:117744311:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TTTTCTGTGCCTATTTCATTTCAGCTTCCCCGAGGCCAGAGATTTCCCGATACATATCATCATGTCTTACATTCACTGCTTCTTGCAATAATTCCCCATGT',\n",
       "  'binding_left': 'GCCAGAGATTTCCCGAT',\n",
       "  'Tm_left': 50.675890815048604,\n",
       "  'binding_right': 'ACATATCATCATGTCTTACATTC',\n",
       "  'Tm_right': 45.11631637403622,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'KCNQ5',\n",
       "  'id': 'chromosome:GRCh37:6:73713630:73713730:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GTTTTCTCCTTGTCTTTGGTTGCTTGATTTTGTCAGTGTTTTCTACCATCACTGAGCACACAAAATTGGCCTCAAGTTGCCTCTTGATCCTGGTAAGTGAA',\n",
       "  'binding_left': 'GTCAGTGTTTTCTACCATCA',\n",
       "  'Tm_left': 50.67075109737817,\n",
       "  'binding_right': 'CTGAGCACACAAAATTGGCC',\n",
       "  'Tm_right': 46.23135800753539,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'PCNX2',\n",
       "  'id': 'chromosome:GRCh37:1:233275425:233275525:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CAAATCATTATGAAGATAAACCATTAACCACTTACTGGGTACCAAGTCTCTGGTGATTGCTTATTAAAAATGCATCAATAGTGAGGGCATTCAAAATTAGC',\n",
       "  'binding_left': 'ACTGGGTACCAAGTCTCT',\n",
       "  'Tm_left': 49.86710724240589,\n",
       "  'binding_right': 'GGTGATTGCTTATTAAAAATGCATCA',\n",
       "  'Tm_right': 45.540004881481025,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'IPO8',\n",
       "  'id': 'chromosome:GRCh37:12:30837332:30837432:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTGGTGGAGGTTCTCGATCTGGCCAGTATTGTGTCACCATGTTCTTCAGGCAAATGGCAGCTGAGTTTGAGAAAAAATGCTAATGTAGCAGTACCAAAAAG',\n",
       "  'binding_left': 'GTCACCATGTTCTTCAGGC',\n",
       "  'Tm_left': 54.51273701683186,\n",
       "  'binding_right': 'AAATGGCAGCTGAGTTTGAGA',\n",
       "  'Tm_right': 46.47612097639063,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'SMS',\n",
       "  'id': 'chromosome:GRCh37:X:21989970:21990070:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GCCCCGATACAATTCTTAAGCTTAAAATTATATTTTCCTTTGTAGCTTTGCAATTTGAGAATTTACCCACATGGATTGGTGTTGCTGGACCTTCAGAGTT',\n",
       "  'binding_left': 'ATATTTTCCTTTGTAGCTTTG',\n",
       "  'Tm_left': 45.3434233882943,\n",
       "  'binding_right': 'CAATTTGAGAATTTACCCACATGG',\n",
       "  'Tm_right': 46.35964217902034,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'RDH8',\n",
       "  'id': 'chromosome:GRCh37:19:10132172:10132272:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTCCCTGGGTCCCAGCGCTCAGGGCCTCCATTCTGCAGGCCATTGTCAACATCATCAGCTCGACTCGACCACCCCTGCGCCGACAGACCAACATCCGCTAC',\n",
       "  'binding_left': 'TCTGCAGGCCATTGTCAACA',\n",
       "  'Tm_left': 53.67817450218888,\n",
       "  'binding_right': 'TCATCAGCTCGACTCGACCA',\n",
       "  'Tm_right': 53.80692976877356,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'IRAK3',\n",
       "  'id': 'chromosome:GRCh37:12:66641479:66641579:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TAAATACTCTTGAAAGTACTCAAGCCAGCTTGTATTTTGCTGAAGATCCTACCACATCACTAAAGTCCTTCAGGTGTCCTTCTCCTCTATTCCTGGAGAAT',\n",
       "  'binding_left': 'TGTATTTTGCTGAAGATCCTA',\n",
       "  'Tm_left': 45.1663000919541,\n",
       "  'binding_right': 'CCACATCACTAAAGTCCTTC',\n",
       "  'Tm_right': 47.54679977007811,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'FBXO42',\n",
       "  'id': 'chromosome:GRCh37:1:16577600:16577700:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GGCCGAGGGGCCTTTGGAGGACATCGCTTTGATGGCTTCCAGACTCCGACACAGGGCACCTGGGGAGACGGCCCCTGCAAGGGCACTGGCCACGTGAGGTG',\n",
       "  'binding_left': 'ATGGCTTCCAGACTCCGACA',\n",
       "  'Tm_left': 55.72407433688517,\n",
       "  'binding_right': 'CAGGGCACCTGGGGAGACGG',\n",
       "  'Tm_right': 59.552781879661836,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'BCAP29;DUS4L-BCAP29',\n",
       "  'id': 'chromosome:GRCh37:7:107254073:107254173:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GGAATTGCTGACTTGCATACTTACAGCATTCCAGTTTTGGTGAATTTTTACGCAAAAGAAGCCACAAAAATGGCAGCATTGGAAAACAGACTGGTTCTAGA',\n",
       "  'binding_left': 'TCCAGTTTTGGTGAATTTTTAC',\n",
       "  'Tm_left': 45.617590348391104,\n",
       "  'binding_right': 'GCAAAAGAAGCCACAAAAATGGCAGC',\n",
       "  'Tm_right': 47.72234870074209,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'RNF213',\n",
       "  'id': 'chromosome:GRCh37:17:78335492:78335592:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTTAGACGCACTCCAGGCCGCTCCTTGGTTTCCTCTCTGCAGGACTGCCCGAGAAGTTCGTGGACATCTTTCAGCAGACTCCTCTGGGCAGGTTTCTTGCC',\n",
       "  'binding_left': 'TCTCTGCAGGACTGCCCG',\n",
       "  'Tm_left': 57.911064609199,\n",
       "  'binding_right': 'AGAAGTTCGTGGACATCTTTCA',\n",
       "  'Tm_right': 48.71116426156368,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'MAFA',\n",
       "  'id': 'chromosome:GRCh37:8:144511652:144511752:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GGTCCCGCTCTTTGGCCAGGCGCCCCACCTCCAGCTTCAGCTGCTCCACCCGGCTCTGGAGTTGGCACTTCTCGCTCTCCAGAATGTGCCGCTGCTGCACC',\n",
       "  'binding_left': 'CAGCTTCAGCTGCTCCACCC',\n",
       "  'Tm_left': 64.24929065484349,\n",
       "  'binding_right': 'GGCTCTGGAGTTGGCACTTC',\n",
       "  'Tm_right': 58.45287691914763,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'AGL',\n",
       "  'id': 'chromosome:GRCh37:1:100347167:100347267:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TTGTGGCTGTATCTAGAACTGCTTTCAGGAATCCCAAGACTTCATTTTACGGCAAGGAAGTGCCTCAAATGTGCATCCCTGGTAATGCAATCTAAAAATTG',\n",
       "  'binding_left': 'TCCCAAGACTTCATTTTACG',\n",
       "  'Tm_left': 45.63103417561234,\n",
       "  'binding_right': 'GCAAGGAAGTGCCTCAAATG',\n",
       "  'Tm_right': 45.870191164932635,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'LYZ',\n",
       "  'id': 'chromosome:GRCh37:12:69743969:69744069:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GGATATTTCAGATCAATAGCCGCTACTGGTGTAATGATGGCAAAACCCCAAGAGCAGTTAATGCCTGTCATTTATCCTGCAGTGGTAAGACAAGCTAATAT',\n",
       "  'binding_left': 'GTAATGATGGCAAAACCCCAA',\n",
       "  'Tm_left': 45.86075753016553,\n",
       "  'binding_right': 'GAGCAGTTAATGCCTGTCAT',\n",
       "  'Tm_right': 50.57468700632563,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'MTMR2',\n",
       "  'id': 'chromosome:GRCh37:11:95656990:95657090:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTCTCCATCCCCTTCGGACGCTGGGCGGGGCCCTGGGCGTCCTGGTTACCCGGACAAGGAGTCCACGCTGGGCGGCCGAGCCGCCGCCGGCTGGGAGCCAA',\n",
       "  'binding_left': 'CTGGGCGTCCTGGTTACCC',\n",
       "  'Tm_left': 62.47879854983432,\n",
       "  'binding_right': 'GGACAAGGAGTCCACGCTGGG',\n",
       "  'Tm_right': 57.13769397801008,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'RBM6',\n",
       "  'id': 'chromosome:GRCh37:3:50091747:50091856:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TGATTTCCTTCTTCCTTACAGTGTAAGGCAAACATTGGTGGGCACCGATCTTCTGCAAGAACCCAAGAGAAGGTGAGTGGCGAAAGTGGTAGCAGTTTTT',\n",
       "  'binding_left': 'AACATTGGTGGGCACCGATC',\n",
       "  'Tm_left': 51.74878934475424,\n",
       "  'binding_right': 'TTCTGCAAGAACCCAAGAGA',\n",
       "  'Tm_right': 45.858872134149635,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'ESR1',\n",
       "  'id': 'chromosome:GRCh37:6:152419906:152420006:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'GAACGTGGTGCCCCTCTATGACCTGCTGCTGGAGATGCTGGACGCCCACCACCTACATGCGCCCACTAGCCGTGGAGGGGCATCCGTGGAGGAGACGGACC',\n",
       "  'binding_left': 'GAGATGCTGGACGCCCACCA',\n",
       "  'Tm_left': 59.512868977677215,\n",
       "  'binding_right': 'CCTACATGCGCCCACTAGCC',\n",
       "  'Tm_right': 61.51358535953534,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'FTL',\n",
       "  'id': 'chromosome:GRCh37:19:49469018:49469119:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'TTCCCGTAGGGCTTCTATTTCGACCGCGATGATGTGGCTCTGGAAGGCGTAGGCCACTTCTTCCGCGAATTGGCCGAGGAGAAGCGCGAGGGCTACGAGCGT',\n",
       "  'binding_left': 'TGTGGCTCTGGAAGGCGTAG',\n",
       "  'Tm_left': 54.07943080742916,\n",
       "  'binding_right': 'GCCACTTCTTCCGCGAATTG',\n",
       "  'Tm_right': 53.721558615446384,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}},\n",
       " {'gene': 'KCNJ15',\n",
       "  'id': 'chromosome:GRCh37:21:39671159:39671259:1',\n",
       "  'molecule': 'dna',\n",
       "  'seq': 'CTCTAAGTGTTTCCAGAGCCTGGCAATGGATGCCATTCACATCGGCATGTGCAGCACCCCCCTGGTGAAGCACACTGCTGGGGCTGGGCTCAAGGCCAACA',\n",
       "  'binding_left': 'GCCATTCACATCGGCATGTG',\n",
       "  'Tm_left': 54.25674554765669,\n",
       "  'binding_right': 'CAGCACCCCCCTGGTGAAGC',\n",
       "  'Tm_right': 64.04783882190026,\n",
       "  'align_num': 0,\n",
       "  'descrip': {}}]"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_sequences_with_blast"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## threshold by blast results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "def thre_by_blast(sequences, ):\n",
    "    non_specific = []\n",
    "    for sequence in sequences:\n",
    "        gene = sequence['gene']\n",
    "        descrip = sequence['descrip']\n",
    "        flag = True\n",
    "        for desc in descrip.values():\n",
    "            if gene.upper() not in desc.upper():\n",
    "                flag = False\n",
    "        if not flag:\n",
    "            non_specific.append(sequence)\n",
    "    \n",
    "    return non_specific"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Tm_left': 48.882851092581916,\n",
      "  'Tm_right': 45.69273727017128,\n",
      "  'align_num': 3,\n",
      "  'binding_left': 'CCACTTTAAATCCCATCATCT',\n",
      "  'binding_right': 'ATGGCATGAGGACCAAACAATT',\n",
      "  'descrip': {1: 'gi|52353333|ref|NM_001005169.1| Homo sapiens olfactory '\n",
      "                 'receptor family 52 subfamily I member 1 (OR52I1), mRNA| pm=1',\n",
      "              2: 'gi|2250017162|ref|NM_001005170.4| Homo sapiens olfactory '\n",
      "                 'receptor family 52 subfamily I member 2 (OR52I2), mRNA| pm=1',\n",
      "              3: 'gi|2234200095|ref|NM_001405760.1| Homo sapiens olfactory '\n",
      "                 'receptor family 52 subfamily I member 2 (OR52I2), mRNA| '\n",
      "                 'pm=1'},\n",
      "  'gene': 'OR52I1',\n",
      "  'id': 'chromosome:GRCh37:11:4616107:4616207:1',\n",
      "  'molecule': 'dna',\n",
      "  'query': '11:4616157..4616157:1',\n",
      "  'seq': 'TGCTAGCTGACCTGTACGTGATCATCCCAGCCACTTTAAATCCCATCATCTATGGCATGAGGACCAAACAATTGCTGGAGGGAATATGGAGTTATCTGATG'},\n",
      " {'Tm_left': 57.96920986402972,\n",
      "  'Tm_right': 49.83280974624722,\n",
      "  'align_num': 11,\n",
      "  'binding_left': 'TGCCCGTCATCGTCTCCAA',\n",
      "  'binding_right': 'CTTCAACTACTTCTACCACCG',\n",
      "  'descrip': {1: 'gi|1653961222|ref|NM_002234.4| Homo sapiens potassium '\n",
      "                 'voltage-gated channel subfamily A member 5 (KCNA5), mRNA| '\n",
      "                 'pm=1',\n",
      "              2: 'gi|2462531786|ref|XM_054371975.1| PREDICTED: Homo sapiens '\n",
      "                 'potassium voltage-gated channel subfamily A member 6 '\n",
      "                 '(KCNA6), transcript variant X3, mRNA| pm=1',\n",
      "              3: 'gi|2462531784|ref|XM_054371974.1| PREDICTED: Homo sapiens '\n",
      "                 'potassium voltage-gated channel subfamily A member 6 '\n",
      "                 '(KCNA6), transcript variant X2, mRNA| pm=1',\n",
      "              4: 'gi|2462531782|ref|XM_054371973.1| PREDICTED: Homo sapiens '\n",
      "                 'potassium voltage-gated channel subfamily A member 6 '\n",
      "                 '(KCNA6), transcript variant X1, mRNA| pm=1',\n",
      "              5: 'gi|2217288914|ref|XM_017019272.2| PREDICTED: Homo sapiens '\n",
      "                 'potassium voltage-gated channel subfamily A member 6 '\n",
      "                 '(KCNA6), transcript variant X3, mRNA| pm=1',\n",
      "              6: 'gi|2217288913|ref|XM_017019271.2| PREDICTED: Homo sapiens '\n",
      "                 'potassium voltage-gated channel subfamily A member 6 '\n",
      "                 '(KCNA6), transcript variant X2, mRNA| pm=1',\n",
      "              7: 'gi|2217288912|ref|XM_017019270.2| PREDICTED: Homo sapiens '\n",
      "                 'potassium voltage-gated channel subfamily A member 6 '\n",
      "                 '(KCNA6), transcript variant X1, mRNA| pm=1',\n",
      "              8: 'gi|2056519540|ref|NM_002235.5| Homo sapiens potassium '\n",
      "                 'voltage-gated channel subfamily A member 6 (KCNA6), '\n",
      "                 'transcript variant 4, mRNA| pm=1',\n",
      "              9: 'gi|2056519539|ref|NR_172524.1| Homo sapiens potassium '\n",
      "                 'voltage-gated channel subfamily A member 6 (KCNA6), '\n",
      "                 'transcript variant 2, non-coding RNA| pm=1',\n",
      "              10: 'gi|2056519538|ref|NR_172525.1| Homo sapiens potassium '\n",
      "                  'voltage-gated channel subfamily A member 6 (KCNA6), '\n",
      "                  'transcript variant 3, non-coding RNA| pm=1',\n",
      "              11: 'gi|2056519537|ref|NR_172523.1| Homo sapiens potassium '\n",
      "                  'voltage-gated channel subfamily A member 6 (KCNA6), '\n",
      "                  'transcript variant 1, non-coding RNA| pm=1'},\n",
      "  'gene': 'KCNA5',\n",
      "  'id': 'chromosome:GRCh37:12:5154816:5154916:1',\n",
      "  'molecule': 'dna',\n",
      "  'query': '12:5154866..5154866:1',\n",
      "  'seq': 'CATCGCCGGGGTCCTCACCATTGCCCTGCCTGTGCCCGTCATCGTCTCCAACTTCAACTACTTCTACCACCGGGAAACGGATCACGAGGAGCCGGCAGTCC'},\n",
      " {'Tm_left': 49.58138954772738,\n",
      "  'Tm_right': 46.65784784500579,\n",
      "  'align_num': 7,\n",
      "  'binding_left': 'TCGGACCTGTCTGTCA',\n",
      "  'binding_right': 'TACAGAGGAATCCAAAACGAATCACC',\n",
      "  'descrip': {1: 'gi|1519241763|ref|NM_006649.4| Homo sapiens UTP14A small '\n",
      "                 'subunit processome component (UTP14A), transcript variant 1, '\n",
      "                 'mRNA| pm=1',\n",
      "              2: 'gi|2217390932|ref|XM_047441791.1| PREDICTED: Homo sapiens '\n",
      "                 'UTP14A small subunit processome component (UTP14A), '\n",
      "                 'transcript variant X2, mRNA '\n",
      "                 '>gi|2462628095|ref|XM_054326403.1| PREDICTED: Homo sapiens '\n",
      "                 'UTP14A small subunit processome component (UTP14A), '\n",
      "                 'transcript variant X2, mRNA| pm=1',\n",
      "              3: 'gi|2217390930|ref|XM_047441790.1| PREDICTED: Homo sapiens '\n",
      "                 'UTP14A small subunit processome component (UTP14A), '\n",
      "                 'transcript variant X1, mRNA '\n",
      "                 '>gi|2462628093|ref|XM_054326402.1| PREDICTED: Homo sapiens '\n",
      "                 'UTP14A small subunit processome component (UTP14A), '\n",
      "                 'transcript variant X1, mRNA| pm=1',\n",
      "              4: 'gi|1890341714|ref|NM_001166221.2| Homo sapiens UTP14A small '\n",
      "                 'subunit processome component (UTP14A), transcript variant 2, '\n",
      "                 'mRNA| pm=1',\n",
      "              5: 'gi|1676452892|ref|NR_036571.3| Homo sapiens ALG11 '\n",
      "                 'alpha-1,2-mannosyltransferase (ALG11), transcript variant B, '\n",
      "                 'non-coding RNA| pm=1',\n",
      "              6: 'gi|1653961346|ref|NM_021645.6| Homo sapiens UTP14C small '\n",
      "                 'subunit processome component (UTP14C), mRNA| pm=1',\n",
      "              7: 'gi|1519312533|ref|NM_001004127.3| Homo sapiens ALG11 '\n",
      "                 'alpha-1,2-mannosyltransferase (ALG11), transcript variant A, '\n",
      "                 'mRNA| pm=1'},\n",
      "  'gene': 'UTP14A',\n",
      "  'id': 'chromosome:GRCh37:X:129063466:129063566:1',\n",
      "  'molecule': 'dna',\n",
      "  'query': 'X:129063516..129063516:1',\n",
      "  'seq': 'AAGCAGAAGACGTGGGCTACCGGTCTTCCTCAAGGTCGGACCTGTCTGTCATACAGAGGAATCCAAAACGAATCACCACACGTCACAAAAAACAGCTGAAG'},\n",
      " {'Tm_left': 58.18263250525655,\n",
      "  'Tm_right': 66.10017392809755,\n",
      "  'align_num': 22,\n",
      "  'binding_left': 'CTTCATCATCCTCTCCAGCTT',\n",
      "  'binding_right': 'CCAGGCTTGGCCCCGGGAG',\n",
      "  'descrip': {1: 'gi|1280409834|ref|NM_001358416.1| Homo sapiens ankyrin '\n",
      "                 'repeat domain 30B like (ANKRD30BL), transcript variant 3, '\n",
      "                 'mRNA| pm=-1',\n",
      "              2: 'gi|2462574894|ref|XR_008486443.1| PREDICTED: Homo sapiens '\n",
      "                 'ankyrin repeat domain 30B like (ANKRD30BL), transcript '\n",
      "                 'variant X2, misc_RNA| pm=-1',\n",
      "              3: 'gi|2462574892|ref|XM_054342831.1| PREDICTED: Homo sapiens '\n",
      "                 'ankyrin repeat domain 30B like (ANKRD30BL), transcript '\n",
      "                 'variant X1, mRNA| pm=-1',\n",
      "              4: 'gi|2217329296|ref|XR_002959312.2| PREDICTED: Homo sapiens '\n",
      "                 'ankyrin repeat domain 30B like (ANKRD30BL), transcript '\n",
      "                 'variant X2, misc_RNA| pm=-1',\n",
      "              5: 'gi|2217329294|ref|XM_047444913.1| PREDICTED: Homo sapiens '\n",
      "                 'ankyrin repeat domain 30B like (ANKRD30BL), transcript '\n",
      "                 'variant X1, mRNA| pm=-1',\n",
      "              6: 'gi|1890608253|ref|NR_027019.3| Homo sapiens ankyrin repeat '\n",
      "                 'domain 30B like (ANKRD30BL), transcript variant 1, '\n",
      "                 'non-coding RNA| pm=-1',\n",
      "              7: 'gi|1890527645|ref|NR_164000.2| Homo sapiens ankyrin repeat '\n",
      "                 'domain 30B like (ANKRD30BL), transcript variant 6, '\n",
      "                 'non-coding RNA| pm=-1',\n",
      "              8: 'gi|1890351095|ref|NR_163999.2| Homo sapiens ankyrin repeat '\n",
      "                 'domain 30B like (ANKRD30BL), transcript variant 5, '\n",
      "                 'non-coding RNA| pm=-1',\n",
      "              9: 'gi|1034672164|ref|XR_001746458.1| PREDICTED: Homo sapiens '\n",
      "                 'ankyrin repeat domain 30B pseudogene (LOC100132004), '\n",
      "                 'transcript variant X3, misc_RNA| pm=-1',\n",
      "              10: 'gi|2462627822|ref|XM_054364454.1| PREDICTED: Homo sapiens '\n",
      "                  'putative ankyrin repeat domain-containing protein 30B-like '\n",
      "                  '(LOC128966581), transcript variant X4, mRNA| pm=-1',\n",
      "              11: 'gi|2462627820|ref|XM_054364453.1| PREDICTED: Homo sapiens '\n",
      "                  'putative ankyrin repeat domain-containing protein 30B-like '\n",
      "                  '(LOC128966581), transcript variant X3, mRNA| pm=-1',\n",
      "              12: 'gi|2462627818|ref|XM_054364452.1| PREDICTED: Homo sapiens '\n",
      "                  'putative ankyrin repeat domain-containing protein 30B-like '\n",
      "                  '(LOC128966581), transcript variant X2, mRNA| pm=-1',\n",
      "              13: 'gi|2462627816|ref|XM_054364451.1| PREDICTED: Homo sapiens '\n",
      "                  'putative ankyrin repeat domain-containing protein 30B-like '\n",
      "                  '(LOC128966581), transcript variant X1, mRNA| pm=-1',\n",
      "              14: 'gi|2462627777|ref|XR_008488144.1| PREDICTED: Homo sapiens '\n",
      "                  'ankyrin repeat domain 30B pseudogene (LOC100132004), '\n",
      "                  'transcript variant X5, misc_RNA| pm=-1',\n",
      "              15: 'gi|2462627776|ref|XR_008488143.1| PREDICTED: Homo sapiens '\n",
      "                  'ankyrin repeat domain 30B pseudogene (LOC100132004), '\n",
      "                  'transcript variant X1, misc_RNA| pm=-1',\n",
      "              16: 'gi|2217384592|ref|XR_001746679.2| PREDICTED: Homo sapiens '\n",
      "                  'ankyrin repeat domain 30B pseudogene (LOC100132154), '\n",
      "                  'transcript variant X2, misc_RNA '\n",
      "                  '>gi|2462627778|ref|XR_008488145.1| PREDICTED: Homo sapiens '\n",
      "                  'ankyrin repeat domain 30B pseudogene (LOC100132004), '\n",
      "                  'transcript variant X4, misc_RNA| pm=-1',\n",
      "              17: 'gi|2217384591|ref|XR_007061544.1| PREDICTED: Homo sapiens '\n",
      "                  'ankyrin repeat domain 30B pseudogene (LOC100132154), '\n",
      "                  'transcript variant X1, misc_RNA| pm=-1',\n",
      "              18: 'gi|2217383851|ref|XR_007061505.1| PREDICTED: Homo sapiens '\n",
      "                  'ankyrin repeat domain 30B pseudogene (LOC100132004), '\n",
      "                  'transcript variant X4, misc_RNA| pm=-1',\n",
      "              19: 'gi|2217383850|ref|XR_007061504.1| PREDICTED: Homo sapiens '\n",
      "                  'ankyrin repeat domain 30B pseudogene (LOC100132004), '\n",
      "                  'transcript variant X2, misc_RNA| pm=-1',\n",
      "              20: 'gi|2217383849|ref|XR_001746457.2| PREDICTED: Homo sapiens '\n",
      "                  'ankyrin repeat domain 30B pseudogene (LOC100132004), '\n",
      "                  'transcript variant X1, misc_RNA| pm=-1',\n",
      "              21: 'gi|2199215665|ref|NR_027417.2| Homo sapiens ankyrin repeat '\n",
      "                  'domain 30B pseudogene (LOC644669), non-coding RNA| pm=-1',\n",
      "              22: 'gi|223941873|ref|NR_026916.1| Homo sapiens ankyrin repeat '\n",
      "                  'domain 30B pseudogene 2 (ANKRD30BP2), non-coding RNA| '\n",
      "                  'pm=-1'},\n",
      "  'gene': 'ANKRD30BL',\n",
      "  'id': 'chromosome:GRCh37:2:132919072:132919172:1',\n",
      "  'molecule': 'dna',\n",
      "  'query': '2:132919122..132919122:1',\n",
      "  'seq': 'ATCTCTTATGTTCAGGTCCATTGTCGTCTTCTTCATCATCCTCTCCAGCTTCCAGGCTTGGCCCCGGGAGGCAGCTTTGTGGATCTTCCTGAGATCCCCAT'},\n",
      " {'Tm_left': 48.91346404943357,\n",
      "  'Tm_right': 53.48709847441535,\n",
      "  'align_num': 28,\n",
      "  'binding_left': 'GCTTAGGGAACAAACTGTCCA',\n",
      "  'binding_right': 'TGAGGTCACCACAGTAGGC',\n",
      "  'descrip': {1: 'gi|1519241854|ref|NM_001320047.2| Homo sapiens FIGNL1 '\n",
      "                 'interacting regulator of recombination and mitosis (FIRRM), '\n",
      "                 'transcript variant 2, mRNA| pm=1',\n",
      "              2: 'gi|1493556043|ref|NR_159440.1| Homo sapiens FIGNL1 '\n",
      "                 'interacting regulator of recombination and mitosis (FIRRM), '\n",
      "                 'transcript variant 8, non-coding RNA| pm=1',\n",
      "              3: 'gi|1493556023|ref|NM_001366769.1| Homo sapiens FIGNL1 '\n",
      "                 'interacting regulator of recombination and mitosis (FIRRM), '\n",
      "                 'transcript variant 10, mRNA| pm=1',\n",
      "              4: 'gi|1493556019|ref|NM_001366772.1| Homo sapiens FIGNL1 '\n",
      "                 'interacting regulator of recombination and mitosis (FIRRM), '\n",
      "                 'transcript variant 11, mRNA| pm=1',\n",
      "              5: 'gi|1493555999|ref|NM_001366770.1| Homo sapiens FIGNL1 '\n",
      "                 'interacting regulator of recombination and mitosis (FIRRM), '\n",
      "                 'transcript variant 13, mRNA| pm=1',\n",
      "              6: 'gi|1493555996|ref|NM_001366768.1| Homo sapiens FIGNL1 '\n",
      "                 'interacting regulator of recombination and mitosis (FIRRM), '\n",
      "                 'transcript variant 7, mRNA| pm=1',\n",
      "              7: 'gi|1493555994|ref|NM_001366773.1| Homo sapiens FIGNL1 '\n",
      "                 'interacting regulator of recombination and mitosis (FIRRM), '\n",
      "                 'transcript variant 12, mRNA| pm=1',\n",
      "              8: 'gi|1493555990|ref|NM_001366771.1| Homo sapiens FIGNL1 '\n",
      "                 'interacting regulator of recombination and mitosis (FIRRM), '\n",
      "                 'transcript variant 9, mRNA| pm=1',\n",
      "              9: 'gi|2462511201|ref|XR_008486103.1| PREDICTED: Homo sapiens '\n",
      "                 'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                 '(FIRRM), transcript variant X9, misc_RNA| pm=1',\n",
      "              10: 'gi|2462511199|ref|XM_054337573.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X8, mRNA| pm=1',\n",
      "              11: 'gi|2462511197|ref|XM_054337572.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X7, mRNA| pm=1',\n",
      "              12: 'gi|2462511195|ref|XM_054337571.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X6, mRNA| pm=1',\n",
      "              13: 'gi|2462511193|ref|XM_054337570.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X5, mRNA| pm=1',\n",
      "              14: 'gi|2462511191|ref|XM_054337569.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X3, mRNA| pm=1',\n",
      "              15: 'gi|2462511189|ref|XM_054337568.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X2, mRNA| pm=1',\n",
      "              16: 'gi|2462511187|ref|XM_054337567.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X1, mRNA| pm=1',\n",
      "              17: 'gi|2217269039|ref|XM_047424771.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X10, mRNA '\n",
      "                  '>gi|2462511202|ref|XM_054337574.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X10, mRNA| pm=1',\n",
      "              18: 'gi|2217269038|ref|XR_921872.2| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X9, misc_RNA| pm=1',\n",
      "              19: 'gi|2217269036|ref|XM_047424770.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X8, mRNA| pm=1',\n",
      "              20: 'gi|2217269035|ref|XM_017001723.2| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X7, mRNA| pm=1',\n",
      "              21: 'gi|2217269033|ref|XM_047424762.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X6, mRNA| pm=1',\n",
      "              22: 'gi|2217269031|ref|XM_047424760.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X5, mRNA| pm=1',\n",
      "              23: 'gi|2217269029|ref|XM_047424759.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X4, mRNA| pm=1',\n",
      "              24: 'gi|2217269027|ref|XM_047424752.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X3, mRNA| pm=1',\n",
      "              25: 'gi|2217269025|ref|XM_047424751.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X2, mRNA| pm=1',\n",
      "              26: 'gi|2217269023|ref|XM_047424743.1| PREDICTED: Homo sapiens '\n",
      "                  'FIGNL1 interacting regulator of recombination and mitosis '\n",
      "                  '(FIRRM), transcript variant X1, mRNA| pm=1',\n",
      "              27: 'gi|1890284637|ref|NM_001363739.2| Homo sapiens FIGNL1 '\n",
      "                  'interacting regulator of recombination and mitosis (FIRRM), '\n",
      "                  'transcript variant 6, mRNA| pm=1',\n",
      "              28: 'gi|1889568737|ref|NM_001320048.2| Homo sapiens FIGNL1 '\n",
      "                  'interacting regulator of recombination and mitosis (FIRRM), '\n",
      "                  'transcript variant 3, mRNA| pm=1'},\n",
      "  'gene': 'C1orf112',\n",
      "  'id': 'chromosome:GRCh37:1:169806124:169806224:1',\n",
      "  'molecule': 'dna',\n",
      "  'query': '1:169806174..169806174:1',\n",
      "  'seq': 'ACATATTTCCTTCCAGGCGTTACCTCCTGAGCTTAGGGAACAAACTGTCCATGAGGTCACCACAGTAGGCACTGCAGAATGCAGGAAATGGCTGAGCAGGA'},\n",
      " {'Tm_left': 45.251165115483275,\n",
      "  'Tm_right': 47.72234870074209,\n",
      "  'align_num': 4,\n",
      "  'binding_left': 'ATTCCAGTTTTGGTGAATTTTTAA',\n",
      "  'binding_right': 'GCAAAAGAAGCCACAAAAATGGCAGC',\n",
      "  'descrip': {1: 'gi|1706853639|ref|NR_163940.2| Homo sapiens DUS4L-BCAP29 '\n",
      "                 'readthrough (DUS4L-BCAP29), transcript variant 6, non-coding '\n",
      "                 'RNA| pm=1',\n",
      "              2: 'gi|1706853634|ref|NM_001371364.2| Homo sapiens DUS4L-BCAP29 '\n",
      "                 'readthrough (DUS4L-BCAP29), transcript variant 1, mRNA| pm=1',\n",
      "              3: 'gi|1701973602|ref|NR_163926.1| Homo sapiens B cell receptor '\n",
      "                 'associated protein 29 (BCAP29), transcript variant 11, '\n",
      "                 'non-coding RNA| pm=1',\n",
      "              4: 'gi|1700611630|ref|NM_001008405.4| Homo sapiens B cell '\n",
      "                 'receptor associated protein 29 (BCAP29), transcript variant '\n",
      "                 '1, mRNA| pm=1'},\n",
      "  'gene': 'BCAP29;DUS4L-BCAP29',\n",
      "  'id': 'chromosome:GRCh37:7:107254073:107254173:1',\n",
      "  'molecule': 'dna',\n",
      "  'query': '7:107254123..107254123:1',\n",
      "  'seq': 'GGAATTGCTGACTTGCATACTTACAGCATTCCAGTTTTGGTGAATTTTTAAGCAAAAGAAGCCACAAAAATGGCAGCATTGGAAAACAGACTGGTTCTAGA'}]\n"
     ]
    }
   ],
   "source": [
    "non_specific = thre_by_blast(sequences=sequences_with_blast)\n",
    "pprint(non_specific)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Tm_left': 67.81069581147506,\n",
      "  'Tm_right': 57.85155207521393,\n",
      "  'align_num': 5,\n",
      "  'binding_left': 'GGGCAGATGCCGGGCCCTGG',\n",
      "  'binding_right': 'GGAGGCAGAGCTGACAGGGG',\n",
      "  'descrip': {1: 'gi|1676317206|ref|NM_182537.3| Homo sapiens '\n",
      "                 '5-hydroxytryptamine receptor 3D (HTR3D), transcript variant '\n",
      "                 '2, mRNA| pm=1',\n",
      "              2: 'gi|1675144830|ref|NM_001163646.2| Homo sapiens '\n",
      "                 '5-hydroxytryptamine receptor 3D (HTR3D), transcript variant '\n",
      "                 '3, mRNA| pm=1',\n",
      "              3: 'gi|2462587945|ref|XM_054345628.1| PREDICTED: Homo sapiens '\n",
      "                 '5-hydroxytryptamine receptor 3D (HTR3D), transcript variant '\n",
      "                 'X1, mRNA| pm=1',\n",
      "              4: 'gi|2287254618|ref|NM_001410851.1| Homo sapiens '\n",
      "                 '5-hydroxytryptamine receptor 3D (HTR3D), transcript variant '\n",
      "                 '4, mRNA| pm=1',\n",
      "              5: 'gi|223468704|ref|NM_001145143.1| Homo sapiens '\n",
      "                 '5-hydroxytryptamine receptor 3D (HTR3D), transcript variant '\n",
      "                 '1, mRNA| pm=1'},\n",
      "  'gene': 'HTR3E',\n",
      "  'id': 'chromosome:GRCh37:3:183824247:183824347:1',\n",
      "  'molecule': 'dna',\n",
      "  'seq': 'TGTAGGTGTGAAGGAGCCAGAGGTATCAGCAGGGCAGATGCCGGGCCCTGGGGAGGCAGAGCTGACAGGGGGCTCAGAATGGACAAGGGCCCAGCGGGAAC'}]\n"
     ]
    }
   ],
   "source": [
    "non_specific = thre_by_blast(sequences=mut_sequences_with_blast)\n",
    "pprint(non_specific)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "probe_designer",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
